{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview and Notes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What are interesting scenarios for exploring correlations?\n",
    "* Expression-dependency correlations for dependency genes that are druggable. For these, the expression correlates are informative for selection of sensitive cell lines/patients. In addition, causally related expression correlates are suggestive of possible resistance mechanisms or secondary targets for a drug combination.\n",
    "* Dependency-dependency correlations for undruggable genes. For these, the correlates are suggestive of substitute targets, or combinations of targets. However, as combinations are considered the causal interpretation becomes more important.\n",
    "* Expression-expression correlations. In this case one idea would be to explore the strength and sign of relationships from the data vs. assertions from curated databases/literature, particularly from perturbational experiments.\n",
    "* Co-expression patterns. Another possibility is to discover patterns of relationships between genes, e.g. that expression of ubiquitin ligases is correlated with the expression of their targets.\n",
    "\n",
    "Another interesting direction would be to use the information in the INDRA database to discover patterns of mechanistic relationships that predict correlations between dependencies of two genes, or between expression of a gene and dependency of another gene. The simplest possible application of this would be to look at pairwise relationships between two genes (could even extract from SIF), and test to see which types of relationships, and signs of relationships predicted, and amount of evidence predicted a) the sign of a correlation and b) the strength of the correlation.\n",
    "\n",
    "If the matrix is made complete, i.e., if entries are included for having no known relationship, we may find that having no relationship is not predictive of a lack of a correlation.\n",
    "\n",
    "Similarly, there will be many mechanistic relationships that yield detectable correlations.\n",
    "\n",
    "One hypothesis that could be tested is whether downstream siblings were more likely to have large correlations to each other than with their ancestors, highlighting that correlations rarely pick up causal relations in this type of data.\n",
    "\n",
    "Could do this with graph convolutional nets?\n",
    "\n",
    "Another application would be to use the data to refine vague relationships extracted from text, or to address issues with assembly, a la distant supervision.\n",
    "\n",
    "Another direction would be to cluster the correlation graph to find likely intermediates.\n",
    "\n",
    "Or, to use the redundancy of certain gene families to explain why certain correlations do *not* show up, i.e., the gene knockout has no effect because it has a redundant copy."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preliminaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2021-04-30 14:52:17] pybel.config - no configuration found, using default sqlite connection sqlite:////Users/johnbachman/.pybel/pybel_0.14.0_cache.db\n",
      "INFO: [2021-04-30 14:52:19] indra.ontology.bio.ontology - Loading INDRA bio ontology from cache at /Users/johnbachman/.indra/bio_ontology/1.12/bio_ontology.pkl\n",
      "WARNING: [2021-04-30 14:52:28] depmap_analysis.network_functions.net_functions - Indra Grounding service not available. Add INDRA_GROUNDING_SERVICE_URL to `indra/config.ini`\n",
      "WARNING: [2021-04-30 14:52:28] util-constructors - DEPRECATION WARNING: This function is being deprecated.\n",
      "WARNING: [2021-04-30 14:52:30] indra_db.databases - Database postgresql://tester:***@indradb-refresh.cwcetxbvbgrf.us-east-1.rds.amazonaws.com/indradb_test is not available: (psycopg2.OperationalError) timeout expired\n",
      "\n",
      "(Background on this error at: http://sqlalche.me/e/e3q8)\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "from os import environ\n",
    "from os.path import join, expanduser\n",
    "from collections import defaultdict, Counter\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "from indra.databases import hgnc_client\n",
    "from indra.tools import assemble_corpus as ac\n",
    "\n",
    "from depmap_analysis.scripts.depmap_script2 import main as run_depmap\n",
    "from depmap_analysis.scripts.depmap_script2 import mito_file\n",
    "from depmap_analysis.preprocessing import get_mitocarta_info\n",
    "from depmap_analysis.network_functions.depmap_network_functions import corr_matrix_to_generator\n",
    "from depmap_analysis.util.statistics import get_z, get_logp, get_n\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "%matplotlib notebook\n",
    "\n",
    "MAX_FLOAT = np.finfo(np.float).max\n",
    "MIN_FLOAT = np.finfo(np.float).min"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load transcript and dependency data and compute correlations\n",
    "\n",
    "\n",
    "Key variables/functions:\n",
    "* `ccle_df`. CCLE expression data, DataFrame.\n",
    "* `crispr_df`. DepMap CRISPR gene effect (dependency) data, DataFrame.\n",
    "* `crispr_corr`\n",
    "* `crispr_z`\n",
    "* `rnai_df`. DepMap combined RNAi data, DataFrame.\n",
    "* `rna_corr`\n",
    "* `rnai_z`\n",
    "* `dep_z`\n",
    "* `dep_df`. CRISPR and RNAi data combined as averaged z-scores, DataFrame.\n",
    "* `dep_corr`. Correlation matrix between dep-dep, dep-expr, and expr-expr (though expr-expr correlations are only defined in this matrix for the genes also included in the dependency data)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#basedir = 'data/20q4'\n",
    "basedir_john = '/Users/johnbachman/Dropbox/1johndata/Knowledge File/Biology/Research/Big Mechanism/datasets/depmap_analysis/depmap/21q1'\n",
    "basedir = environ.get('BASEPATH', basedir_john)\n",
    "figdir = join(expanduser('~'), 'Dropbox/DARPA projects/papers/INDRA paper 2/figures/figure_panels')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mitocarta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mitocarta = pd.read_excel('data/Human.MitoCarta3.0.xls', sheet_name=1)\n",
    "mitogenes = list(mitocarta.Symbol.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cell line metadata\n",
    "\n",
    "Load the cell line metadata which includes mappings between cell line identifiers used by different datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_line_df = pd.read_csv(join(basedir, 'sample_info.csv'))\n",
    "#cell_line_df = pd.read_csv('data/DepMap-2019q1-celllines_v2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DepMap_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCLE_Name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NIHOVCAR3_OVARY</th>\n",
       "      <td>ACH-000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <td>ACH-000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACO2_LARGE_INTESTINE</th>\n",
       "      <td>ACH-000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <td>ACH-000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <td>ACH-000005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             DepMap_ID\n",
       "CCLE_Name                                             \n",
       "NIHOVCAR3_OVARY                             ACH-000001\n",
       "HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE     ACH-000002\n",
       "CACO2_LARGE_INTESTINE                       ACH-000003\n",
       "HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE      ACH-000004\n",
       "HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  ACH-000005"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_line_map = cell_line_df[['DepMap_ID', 'CCLE_Name']]\n",
    "cell_line_map.set_index('CCLE_Name', inplace=True)\n",
    "cell_line_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions for processing RNAi and CRISPR correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Functions from Albert to calculate the inverse log CDF of the normal distribution:\n",
    "(moved to depmap_analysis.util.statistics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_corr(recalculate, data_df, filename):\n",
    "    filepath = join(basedir, filename)\n",
    "    if recalculate:\n",
    "        %time data_corr = data_df.corr()\n",
    "        data_corr.to_hdf('%s.h5' % filepath, filename)\n",
    "    else:\n",
    "        data_corr = pd.read_hdf('%s.h5' % filepath)\n",
    "    return data_corr\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### RNAi data\n",
    "\n",
    "Load combined RNAi data, normalize column names, map to DepMap cell line IDs, and drop duplicate columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.97 s, sys: 203 ms, total: 2.17 s\n",
      "Wall time: 2.18 s\n"
     ]
    }
   ],
   "source": [
    "%time rnai_df = pd.read_csv(join(basedir, 'D2_combined_gene_dep_scores.csv'), index_col=0)\n",
    "rnai_df = rnai_df.transpose()\n",
    "gene_cols = ['%s' % col.split(' ')[0] for col in rnai_df.columns]\n",
    "rnai_df.columns = gene_cols\n",
    "rnai_df = rnai_df.join(cell_line_map)\n",
    "rnai_df = rnai_df.set_index('DepMap_ID')\n",
    "# Drop duplicate columns\n",
    "rnai_df = rnai_df.loc[:,~rnai_df.columns.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>NAT2</th>\n",
       "      <th>ADA</th>\n",
       "      <th>CDH2</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>MED6</th>\n",
       "      <th>NR2E3</th>\n",
       "      <th>NAALAD2</th>\n",
       "      <th>DUXB</th>\n",
       "      <th>PDZK1P1</th>\n",
       "      <th>...</th>\n",
       "      <th>RCE1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>DMTF1</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>KCNE2</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>SCO2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-001270</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.194962</td>\n",
       "      <td>-0.256108</td>\n",
       "      <td>-0.174220</td>\n",
       "      <td>-0.140052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.201644</td>\n",
       "      <td>-0.363670</td>\n",
       "      <td>0.184260</td>\n",
       "      <td>-0.115616</td>\n",
       "      <td>-0.125958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.088853</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.843295</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001000</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.028171</td>\n",
       "      <td>0.100751</td>\n",
       "      <td>-0.456124</td>\n",
       "      <td>-0.174618</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074889</td>\n",
       "      <td>0.152158</td>\n",
       "      <td>0.036011</td>\n",
       "      <td>0.117300</td>\n",
       "      <td>0.101725</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.110628</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.307031</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001001</th>\n",
       "      <td>0.146042</td>\n",
       "      <td>0.102854</td>\n",
       "      <td>0.168839</td>\n",
       "      <td>0.063047</td>\n",
       "      <td>-0.008077</td>\n",
       "      <td>-0.214376</td>\n",
       "      <td>-0.153619</td>\n",
       "      <td>0.133830</td>\n",
       "      <td>0.138673</td>\n",
       "      <td>0.030345</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006735</td>\n",
       "      <td>-0.033385</td>\n",
       "      <td>0.197651</td>\n",
       "      <td>-0.016372</td>\n",
       "      <td>0.077486</td>\n",
       "      <td>0.106165</td>\n",
       "      <td>0.057286</td>\n",
       "      <td>0.025596</td>\n",
       "      <td>-0.413669</td>\n",
       "      <td>0.122669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-002319</th>\n",
       "      <td>-0.190388</td>\n",
       "      <td>0.384106</td>\n",
       "      <td>-0.120700</td>\n",
       "      <td>-0.237251</td>\n",
       "      <td>0.060267</td>\n",
       "      <td>-0.338946</td>\n",
       "      <td>-0.057551</td>\n",
       "      <td>0.134511</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.144463</td>\n",
       "      <td>...</td>\n",
       "      <td>0.209009</td>\n",
       "      <td>-0.156839</td>\n",
       "      <td>-0.155837</td>\n",
       "      <td>-0.001141</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.227968</td>\n",
       "      <td>0.028095</td>\n",
       "      <td>-0.080611</td>\n",
       "      <td>-1.849696</td>\n",
       "      <td>-0.078856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001827</th>\n",
       "      <td>0.907063</td>\n",
       "      <td>0.403192</td>\n",
       "      <td>0.004394</td>\n",
       "      <td>-0.017059</td>\n",
       "      <td>-0.094749</td>\n",
       "      <td>-0.328074</td>\n",
       "      <td>-0.089573</td>\n",
       "      <td>0.362029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.098161</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.137465</td>\n",
       "      <td>-1.037848</td>\n",
       "      <td>-0.261262</td>\n",
       "      <td>-0.228016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.088744</td>\n",
       "      <td>0.159467</td>\n",
       "      <td>0.014071</td>\n",
       "      <td>-0.414154</td>\n",
       "      <td>0.032661</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17309 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                A1BG      NAT2       ADA      CDH2      AKT3      MED6  \\\n",
       "DepMap_ID                                                                \n",
       "ACH-001270       NaN       NaN       NaN -0.194962 -0.256108 -0.174220   \n",
       "ACH-001000       NaN       NaN       NaN -0.028171  0.100751 -0.456124   \n",
       "ACH-001001  0.146042  0.102854  0.168839  0.063047 -0.008077 -0.214376   \n",
       "ACH-002319 -0.190388  0.384106 -0.120700 -0.237251  0.060267 -0.338946   \n",
       "ACH-001827  0.907063  0.403192  0.004394 -0.017059 -0.094749 -0.328074   \n",
       "\n",
       "               NR2E3   NAALAD2      DUXB   PDZK1P1  ...      RCE1   HNRNPDL  \\\n",
       "DepMap_ID                                           ...                       \n",
       "ACH-001270 -0.140052       NaN       NaN       NaN  ... -0.201644 -0.363670   \n",
       "ACH-001000 -0.174618       NaN       NaN       NaN  ...  0.074889  0.152158   \n",
       "ACH-001001 -0.153619  0.133830  0.138673  0.030345  ...  0.006735 -0.033385   \n",
       "ACH-002319 -0.057551  0.134511       NaN  0.144463  ...  0.209009 -0.156839   \n",
       "ACH-001827 -0.089573  0.362029       NaN -0.098161  ... -0.137465 -1.037848   \n",
       "\n",
       "               DMTF1    PPP4R1      CDH1   SLC12A6     KCNE2     DGCR2  \\\n",
       "DepMap_ID                                                                \n",
       "ACH-001270  0.184260 -0.115616 -0.125958       NaN  0.088853       NaN   \n",
       "ACH-001000  0.036011  0.117300  0.101725       NaN -0.110628       NaN   \n",
       "ACH-001001  0.197651 -0.016372  0.077486  0.106165  0.057286  0.025596   \n",
       "ACH-002319 -0.155837 -0.001141       NaN  0.227968  0.028095 -0.080611   \n",
       "ACH-001827 -0.261262 -0.228016       NaN  0.088744  0.159467  0.014071   \n",
       "\n",
       "            CASP8AP2      SCO2  \n",
       "DepMap_ID                       \n",
       "ACH-001270 -0.843295       NaN  \n",
       "ACH-001000 -0.307031       NaN  \n",
       "ACH-001001 -0.413669  0.122669  \n",
       "ACH-002319 -1.849696 -0.078856  \n",
       "ACH-001827 -0.414154  0.032661  \n",
       "\n",
       "[5 rows x 17309 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnai_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Calculate RNAi correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2021-04-30 15:17:18] numexpr.utils - Note: NumExpr detected 12 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO: [2021-04-30 15:17:18] numexpr.utils - NumExpr defaulting to 8 threads.\n"
     ]
    }
   ],
   "source": [
    "recalculate = False\n",
    "\n",
    "rnai_corr = get_corr(recalculate, rnai_df, 'rnai_correlations')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Calculate RNAi sample sizes (number of non-missing values in pairs of columns):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>NAT2</th>\n",
       "      <th>ADA</th>\n",
       "      <th>CDH2</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>MED6</th>\n",
       "      <th>NR2E3</th>\n",
       "      <th>NAALAD2</th>\n",
       "      <th>DUXB</th>\n",
       "      <th>PDZK1P1</th>\n",
       "      <th>...</th>\n",
       "      <th>RCE1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>DMTF1</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>KCNE2</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>SCO2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.053170</td>\n",
       "      <td>-0.057689</td>\n",
       "      <td>0.053209</td>\n",
       "      <td>-0.019174</td>\n",
       "      <td>0.000930</td>\n",
       "      <td>0.089174</td>\n",
       "      <td>0.053059</td>\n",
       "      <td>0.036935</td>\n",
       "      <td>-0.002463</td>\n",
       "      <td>...</td>\n",
       "      <td>0.119704</td>\n",
       "      <td>-0.146314</td>\n",
       "      <td>-0.024557</td>\n",
       "      <td>0.038307</td>\n",
       "      <td>-0.023972</td>\n",
       "      <td>0.017129</td>\n",
       "      <td>0.024641</td>\n",
       "      <td>-0.047776</td>\n",
       "      <td>0.023925</td>\n",
       "      <td>0.100400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAT2</th>\n",
       "      <td>-0.053170</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.051519</td>\n",
       "      <td>0.002823</td>\n",
       "      <td>-0.034266</td>\n",
       "      <td>0.030885</td>\n",
       "      <td>0.005237</td>\n",
       "      <td>-0.039173</td>\n",
       "      <td>-0.017945</td>\n",
       "      <td>-0.106155</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035899</td>\n",
       "      <td>-0.051940</td>\n",
       "      <td>0.008295</td>\n",
       "      <td>0.008854</td>\n",
       "      <td>-0.159054</td>\n",
       "      <td>-0.007945</td>\n",
       "      <td>0.054352</td>\n",
       "      <td>-0.058752</td>\n",
       "      <td>-0.094724</td>\n",
       "      <td>-0.032338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADA</th>\n",
       "      <td>-0.057689</td>\n",
       "      <td>0.051519</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.028305</td>\n",
       "      <td>-0.038091</td>\n",
       "      <td>0.039840</td>\n",
       "      <td>-0.007825</td>\n",
       "      <td>-0.037555</td>\n",
       "      <td>0.000769</td>\n",
       "      <td>0.016227</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043982</td>\n",
       "      <td>0.010465</td>\n",
       "      <td>0.056808</td>\n",
       "      <td>0.073861</td>\n",
       "      <td>-0.017691</td>\n",
       "      <td>-0.020626</td>\n",
       "      <td>-0.047993</td>\n",
       "      <td>0.106098</td>\n",
       "      <td>0.018621</td>\n",
       "      <td>-0.010055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH2</th>\n",
       "      <td>0.053209</td>\n",
       "      <td>0.002823</td>\n",
       "      <td>-0.028305</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.050971</td>\n",
       "      <td>-0.066889</td>\n",
       "      <td>-0.003648</td>\n",
       "      <td>0.016123</td>\n",
       "      <td>0.022661</td>\n",
       "      <td>-0.022969</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.037606</td>\n",
       "      <td>-0.131892</td>\n",
       "      <td>0.097606</td>\n",
       "      <td>-0.076574</td>\n",
       "      <td>-0.003035</td>\n",
       "      <td>-0.010893</td>\n",
       "      <td>0.007762</td>\n",
       "      <td>-0.104119</td>\n",
       "      <td>-0.081411</td>\n",
       "      <td>-0.022924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT3</th>\n",
       "      <td>-0.019174</td>\n",
       "      <td>-0.034266</td>\n",
       "      <td>-0.038091</td>\n",
       "      <td>0.050971</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.022616</td>\n",
       "      <td>-0.081685</td>\n",
       "      <td>0.052951</td>\n",
       "      <td>0.072240</td>\n",
       "      <td>-0.037919</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.091548</td>\n",
       "      <td>0.027137</td>\n",
       "      <td>0.041509</td>\n",
       "      <td>-0.064823</td>\n",
       "      <td>0.029014</td>\n",
       "      <td>-0.086059</td>\n",
       "      <td>-0.038459</td>\n",
       "      <td>-0.090081</td>\n",
       "      <td>-0.007495</td>\n",
       "      <td>-0.020063</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17309 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A1BG      NAT2       ADA      CDH2      AKT3      MED6     NR2E3  \\\n",
       "A1BG  1.000000 -0.053170 -0.057689  0.053209 -0.019174  0.000930  0.089174   \n",
       "NAT2 -0.053170  1.000000  0.051519  0.002823 -0.034266  0.030885  0.005237   \n",
       "ADA  -0.057689  0.051519  1.000000 -0.028305 -0.038091  0.039840 -0.007825   \n",
       "CDH2  0.053209  0.002823 -0.028305  1.000000  0.050971 -0.066889 -0.003648   \n",
       "AKT3 -0.019174 -0.034266 -0.038091  0.050971  1.000000 -0.022616 -0.081685   \n",
       "\n",
       "       NAALAD2      DUXB   PDZK1P1  ...      RCE1   HNRNPDL     DMTF1  \\\n",
       "A1BG  0.053059  0.036935 -0.002463  ...  0.119704 -0.146314 -0.024557   \n",
       "NAT2 -0.039173 -0.017945 -0.106155  ...  0.035899 -0.051940  0.008295   \n",
       "ADA  -0.037555  0.000769  0.016227  ...  0.043982  0.010465  0.056808   \n",
       "CDH2  0.016123  0.022661 -0.022969  ... -0.037606 -0.131892  0.097606   \n",
       "AKT3  0.052951  0.072240 -0.037919  ... -0.091548  0.027137  0.041509   \n",
       "\n",
       "        PPP4R1      CDH1   SLC12A6     KCNE2     DGCR2  CASP8AP2      SCO2  \n",
       "A1BG  0.038307 -0.023972  0.017129  0.024641 -0.047776  0.023925  0.100400  \n",
       "NAT2  0.008854 -0.159054 -0.007945  0.054352 -0.058752 -0.094724 -0.032338  \n",
       "ADA   0.073861 -0.017691 -0.020626 -0.047993  0.106098  0.018621 -0.010055  \n",
       "CDH2 -0.076574 -0.003035 -0.010893  0.007762 -0.104119 -0.081411 -0.022924  \n",
       "AKT3 -0.064823  0.029014 -0.086059 -0.038459 -0.090081 -0.007495 -0.020063  \n",
       "\n",
       "[5 rows x 17309 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnai_corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 sec\n"
     ]
    }
   ],
   "source": [
    "recalculate = False\n",
    "\n",
    "rnai_n = get_n(recalculate, rnai_df, join(basedir, 'rnai_n'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>NAT2</th>\n",
       "      <th>ADA</th>\n",
       "      <th>CDH2</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>MED6</th>\n",
       "      <th>NR2E3</th>\n",
       "      <th>NAALAD2</th>\n",
       "      <th>DUXB</th>\n",
       "      <th>PDZK1P1</th>\n",
       "      <th>...</th>\n",
       "      <th>RCE1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>DMTF1</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>KCNE2</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>SCO2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>712.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>...</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>507.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAT2</th>\n",
       "      <td>547.0</td>\n",
       "      <td>712.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>...</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>507.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADA</th>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>712.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>...</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>507.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH2</th>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>712.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>708.0</td>\n",
       "      <td>708.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>...</td>\n",
       "      <td>710.0</td>\n",
       "      <td>707.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>670.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>708.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>547.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT3</th>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>712.0</td>\n",
       "      <td>708.0</td>\n",
       "      <td>708.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>...</td>\n",
       "      <td>710.0</td>\n",
       "      <td>707.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>670.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>708.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>710.0</td>\n",
       "      <td>547.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17309 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       A1BG   NAT2    ADA   CDH2   AKT3   MED6  NR2E3  NAALAD2   DUXB  \\\n",
       "A1BG  712.0  547.0  547.0  547.0  547.0  547.0  547.0    547.0  501.0   \n",
       "NAT2  547.0  712.0  547.0  547.0  547.0  547.0  547.0    547.0  501.0   \n",
       "ADA   547.0  547.0  712.0  547.0  547.0  547.0  547.0    547.0  501.0   \n",
       "CDH2  547.0  547.0  547.0  712.0  710.0  708.0  708.0    547.0  501.0   \n",
       "AKT3  547.0  547.0  547.0  710.0  712.0  708.0  708.0    547.0  501.0   \n",
       "\n",
       "      PDZK1P1  ...   RCE1  HNRNPDL  DMTF1  PPP4R1   CDH1  SLC12A6  KCNE2  \\\n",
       "A1BG    547.0  ...  547.0    547.0  547.0   547.0  507.0    547.0  547.0   \n",
       "NAT2    547.0  ...  547.0    547.0  547.0   547.0  507.0    547.0  547.0   \n",
       "ADA     547.0  ...  547.0    547.0  547.0   547.0  507.0    547.0  547.0   \n",
       "CDH2    547.0  ...  710.0    707.0  710.0   710.0  670.0    547.0  708.0   \n",
       "AKT3    547.0  ...  710.0    707.0  710.0   710.0  670.0    547.0  708.0   \n",
       "\n",
       "      DGCR2  CASP8AP2   SCO2  \n",
       "A1BG  547.0     547.0  547.0  \n",
       "NAT2  547.0     547.0  547.0  \n",
       "ADA   547.0     547.0  547.0  \n",
       "CDH2  547.0     710.0  547.0  \n",
       "AKT3  547.0     710.0  547.0  \n",
       "\n",
       "[5 rows x 17309 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnai_n.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Calculate RNAi p-values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.158393144607544 sec\n"
     ]
    }
   ],
   "source": [
    "recalculate = False\n",
    "\n",
    "rnai_logp = get_logp(recalculate, rnai_n, rnai_corr, join(basedir, 'rnai_logp'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>NAT2</th>\n",
       "      <th>ADA</th>\n",
       "      <th>CDH2</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>MED6</th>\n",
       "      <th>NR2E3</th>\n",
       "      <th>NAALAD2</th>\n",
       "      <th>DUXB</th>\n",
       "      <th>PDZK1P1</th>\n",
       "      <th>...</th>\n",
       "      <th>RCE1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>DMTF1</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>KCNE2</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>SCO2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>-inf</td>\n",
       "      <td>-1.539955</td>\n",
       "      <td>-1.726559</td>\n",
       "      <td>-1.541511</td>\n",
       "      <td>-0.423808</td>\n",
       "      <td>-0.017465</td>\n",
       "      <td>-3.294917</td>\n",
       "      <td>-1.535465</td>\n",
       "      <td>-0.893049</td>\n",
       "      <td>-0.046918</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.286829</td>\n",
       "      <td>-7.422472</td>\n",
       "      <td>-0.568168</td>\n",
       "      <td>-0.990980</td>\n",
       "      <td>-0.527266</td>\n",
       "      <td>-0.371983</td>\n",
       "      <td>-0.570510</td>\n",
       "      <td>-1.329329</td>\n",
       "      <td>-0.550615</td>\n",
       "      <td>-3.971862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAT2</th>\n",
       "      <td>-1.539955</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-1.474098</td>\n",
       "      <td>-0.053956</td>\n",
       "      <td>-0.858461</td>\n",
       "      <td>-0.752902</td>\n",
       "      <td>-0.102319</td>\n",
       "      <td>-1.020302</td>\n",
       "      <td>-0.373031</td>\n",
       "      <td>-4.343707</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.911154</td>\n",
       "      <td>-1.490757</td>\n",
       "      <td>-0.166633</td>\n",
       "      <td>-0.178757</td>\n",
       "      <td>-8.035273</td>\n",
       "      <td>-0.159088</td>\n",
       "      <td>-1.587852</td>\n",
       "      <td>-1.771831</td>\n",
       "      <td>-3.621666</td>\n",
       "      <td>-0.797674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADA</th>\n",
       "      <td>-1.726559</td>\n",
       "      <td>-1.474098</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.675583</td>\n",
       "      <td>-0.983717</td>\n",
       "      <td>-1.043109</td>\n",
       "      <td>-0.156524</td>\n",
       "      <td>-0.965797</td>\n",
       "      <td>-0.013796</td>\n",
       "      <td>-0.349666</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.189015</td>\n",
       "      <td>-0.214341</td>\n",
       "      <td>-1.689445</td>\n",
       "      <td>-2.472535</td>\n",
       "      <td>-0.369497</td>\n",
       "      <td>-0.461603</td>\n",
       "      <td>-1.337555</td>\n",
       "      <td>-4.339962</td>\n",
       "      <td>-0.409641</td>\n",
       "      <td>-0.205192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH2</th>\n",
       "      <td>-1.541511</td>\n",
       "      <td>-0.053956</td>\n",
       "      <td>-0.675583</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-1.743591</td>\n",
       "      <td>-2.586298</td>\n",
       "      <td>-0.080327</td>\n",
       "      <td>-0.347111</td>\n",
       "      <td>-0.489658</td>\n",
       "      <td>-0.524363</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.148818</td>\n",
       "      <td>-7.733410</td>\n",
       "      <td>-4.682383</td>\n",
       "      <td>-3.185126</td>\n",
       "      <td>-0.064542</td>\n",
       "      <td>-0.223964</td>\n",
       "      <td>-0.178333</td>\n",
       "      <td>-4.210226</td>\n",
       "      <td>-3.503925</td>\n",
       "      <td>-0.523129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT3</th>\n",
       "      <td>-0.423808</td>\n",
       "      <td>-0.858461</td>\n",
       "      <td>-0.983717</td>\n",
       "      <td>-1.743591</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.601501</td>\n",
       "      <td>-3.514637</td>\n",
       "      <td>-1.531147</td>\n",
       "      <td>-2.241452</td>\n",
       "      <td>-0.977950</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.221340</td>\n",
       "      <td>-0.752339</td>\n",
       "      <td>-1.311764</td>\n",
       "      <td>-2.472910</td>\n",
       "      <td>-0.790988</td>\n",
       "      <td>-3.118282</td>\n",
       "      <td>-1.181474</td>\n",
       "      <td>-3.347257</td>\n",
       "      <td>-0.171997</td>\n",
       "      <td>-0.446845</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17309 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A1BG      NAT2       ADA      CDH2      AKT3      MED6     NR2E3  \\\n",
       "A1BG      -inf -1.539955 -1.726559 -1.541511 -0.423808 -0.017465 -3.294917   \n",
       "NAT2 -1.539955      -inf -1.474098 -0.053956 -0.858461 -0.752902 -0.102319   \n",
       "ADA  -1.726559 -1.474098      -inf -0.675583 -0.983717 -1.043109 -0.156524   \n",
       "CDH2 -1.541511 -0.053956 -0.675583      -inf -1.743591 -2.586298 -0.080327   \n",
       "AKT3 -0.423808 -0.858461 -0.983717 -1.743591      -inf -0.601501 -3.514637   \n",
       "\n",
       "       NAALAD2      DUXB   PDZK1P1  ...      RCE1   HNRNPDL     DMTF1  \\\n",
       "A1BG -1.535465 -0.893049 -0.046918  ... -5.286829 -7.422472 -0.568168   \n",
       "NAT2 -1.020302 -0.373031 -4.343707  ... -0.911154 -1.490757 -0.166633   \n",
       "ADA  -0.965797 -0.013796 -0.349666  ... -1.189015 -0.214341 -1.689445   \n",
       "CDH2 -0.347111 -0.489658 -0.524363  ... -1.148818 -7.733410 -4.682383   \n",
       "AKT3 -1.531147 -2.241452 -0.977950  ... -4.221340 -0.752339 -1.311764   \n",
       "\n",
       "        PPP4R1      CDH1   SLC12A6     KCNE2     DGCR2  CASP8AP2      SCO2  \n",
       "A1BG -0.990980 -0.527266 -0.371983 -0.570510 -1.329329 -0.550615 -3.971862  \n",
       "NAT2 -0.178757 -8.035273 -0.159088 -1.587852 -1.771831 -3.621666 -0.797674  \n",
       "ADA  -2.472535 -0.369497 -0.461603 -1.337555 -4.339962 -0.409641 -0.205192  \n",
       "CDH2 -3.185126 -0.064542 -0.223964 -0.178333 -4.210226 -3.503925 -0.523129  \n",
       "AKT3 -2.472910 -0.790988 -3.118282 -1.181474 -3.347257 -0.171997 -0.446845  \n",
       "\n",
       "[5 rows x 17309 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnai_logp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0007172501778812698"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.exp(rnai_logp['CASP9']['OSCP1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-0.18179857862062887, 0.0007172501778812697)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate explicitly using scipy.stats.pearsonr\n",
    "x, y = rnai_df['CASP9'].values, rnai_df['OSCP1'].values\n",
    "nas = np.logical_or(np.isnan(x), np.isnan(y))\n",
    "corr = stats.pearsonr(x[~nas], y[~nas])\n",
    "corr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Calculate RNAi z-scores:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93.67246508598328 sec\n"
     ]
    }
   ],
   "source": [
    "recalculate = True\n",
    "\n",
    "rnai_z = get_z(recalculate, rnai_logp, rnai_corr, join(basedir, 'rnai_z_log'))\n",
    "#rnai_z_ppf = get_z(recalculate, rnai_logp, rnai_corr, 'rnai_z')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>NAT2</th>\n",
       "      <th>ADA</th>\n",
       "      <th>CDH2</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>MED6</th>\n",
       "      <th>NR2E3</th>\n",
       "      <th>NAALAD2</th>\n",
       "      <th>DUXB</th>\n",
       "      <th>PDZK1P1</th>\n",
       "      <th>...</th>\n",
       "      <th>RCE1</th>\n",
       "      <th>HNRNPDL</th>\n",
       "      <th>DMTF1</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>KCNE2</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>SCO2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>inf</td>\n",
       "      <td>-1.241582</td>\n",
       "      <td>-1.347263</td>\n",
       "      <td>1.242485</td>\n",
       "      <td>-0.447451</td>\n",
       "      <td>0.021701</td>\n",
       "      <td>2.084980</td>\n",
       "      <td>1.238973</td>\n",
       "      <td>0.824940</td>\n",
       "      <td>-0.057476</td>\n",
       "      <td>...</td>\n",
       "      <td>2.803331</td>\n",
       "      <td>-3.432670</td>\n",
       "      <td>-0.573122</td>\n",
       "      <td>0.894204</td>\n",
       "      <td>-0.538522</td>\n",
       "      <td>0.399715</td>\n",
       "      <td>0.575080</td>\n",
       "      <td>-1.115457</td>\n",
       "      <td>0.558365</td>\n",
       "      <td>2.348715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAT2</th>\n",
       "      <td>-1.241582</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.202977</td>\n",
       "      <td>0.065879</td>\n",
       "      <td>-0.799822</td>\n",
       "      <td>0.720857</td>\n",
       "      <td>0.122198</td>\n",
       "      <td>-0.914440</td>\n",
       "      <td>-0.400696</td>\n",
       "      <td>-2.484090</td>\n",
       "      <td>...</td>\n",
       "      <td>0.837948</td>\n",
       "      <td>-1.212811</td>\n",
       "      <td>0.193573</td>\n",
       "      <td>0.206617</td>\n",
       "      <td>-3.595446</td>\n",
       "      <td>-0.185393</td>\n",
       "      <td>1.269216</td>\n",
       "      <td>-1.372135</td>\n",
       "      <td>-2.215320</td>\n",
       "      <td>-0.754789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADA</th>\n",
       "      <td>-1.347263</td>\n",
       "      <td>1.202977</td>\n",
       "      <td>inf</td>\n",
       "      <td>-0.660615</td>\n",
       "      <td>-0.889157</td>\n",
       "      <td>0.930026</td>\n",
       "      <td>-0.182601</td>\n",
       "      <td>-0.876645</td>\n",
       "      <td>0.017173</td>\n",
       "      <td>0.378682</td>\n",
       "      <td>...</td>\n",
       "      <td>1.026786</td>\n",
       "      <td>0.244204</td>\n",
       "      <td>1.326658</td>\n",
       "      <td>1.725871</td>\n",
       "      <td>-0.397387</td>\n",
       "      <td>-0.481344</td>\n",
       "      <td>-1.120533</td>\n",
       "      <td>2.482756</td>\n",
       "      <td>0.434550</td>\n",
       "      <td>-0.234636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH2</th>\n",
       "      <td>1.242485</td>\n",
       "      <td>0.065879</td>\n",
       "      <td>-0.660615</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.356654</td>\n",
       "      <td>-1.778643</td>\n",
       "      <td>-0.096890</td>\n",
       "      <td>0.376256</td>\n",
       "      <td>0.506030</td>\n",
       "      <td>-0.536038</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.000618</td>\n",
       "      <td>-3.516092</td>\n",
       "      <td>2.602417</td>\n",
       "      <td>-2.039772</td>\n",
       "      <td>-0.078417</td>\n",
       "      <td>-0.254197</td>\n",
       "      <td>0.206163</td>\n",
       "      <td>-2.436187</td>\n",
       "      <td>-2.169047</td>\n",
       "      <td>-0.534980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT3</th>\n",
       "      <td>-0.447451</td>\n",
       "      <td>-0.799822</td>\n",
       "      <td>-0.889157</td>\n",
       "      <td>1.356654</td>\n",
       "      <td>inf</td>\n",
       "      <td>-0.600777</td>\n",
       "      <td>-2.173289</td>\n",
       "      <td>1.236460</td>\n",
       "      <td>1.615031</td>\n",
       "      <td>-0.885140</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.440204</td>\n",
       "      <td>0.720425</td>\n",
       "      <td>1.104573</td>\n",
       "      <td>-1.726046</td>\n",
       "      <td>0.749765</td>\n",
       "      <td>-2.011875</td>\n",
       "      <td>-1.021904</td>\n",
       "      <td>-2.106272</td>\n",
       "      <td>-0.199360</td>\n",
       "      <td>-0.468198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17309 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A1BG      NAT2       ADA      CDH2      AKT3      MED6     NR2E3  \\\n",
       "A1BG       inf -1.241582 -1.347263  1.242485 -0.447451  0.021701  2.084980   \n",
       "NAT2 -1.241582       inf  1.202977  0.065879 -0.799822  0.720857  0.122198   \n",
       "ADA  -1.347263  1.202977       inf -0.660615 -0.889157  0.930026 -0.182601   \n",
       "CDH2  1.242485  0.065879 -0.660615       inf  1.356654 -1.778643 -0.096890   \n",
       "AKT3 -0.447451 -0.799822 -0.889157  1.356654       inf -0.600777 -2.173289   \n",
       "\n",
       "       NAALAD2      DUXB   PDZK1P1  ...      RCE1   HNRNPDL     DMTF1  \\\n",
       "A1BG  1.238973  0.824940 -0.057476  ...  2.803331 -3.432670 -0.573122   \n",
       "NAT2 -0.914440 -0.400696 -2.484090  ...  0.837948 -1.212811  0.193573   \n",
       "ADA  -0.876645  0.017173  0.378682  ...  1.026786  0.244204  1.326658   \n",
       "CDH2  0.376256  0.506030 -0.536038  ... -1.000618 -3.516092  2.602417   \n",
       "AKT3  1.236460  1.615031 -0.885140  ... -2.440204  0.720425  1.104573   \n",
       "\n",
       "        PPP4R1      CDH1   SLC12A6     KCNE2     DGCR2  CASP8AP2      SCO2  \n",
       "A1BG  0.894204 -0.538522  0.399715  0.575080 -1.115457  0.558365  2.348715  \n",
       "NAT2  0.206617 -3.595446 -0.185393  1.269216 -1.372135 -2.215320 -0.754789  \n",
       "ADA   1.725871 -0.397387 -0.481344 -1.120533  2.482756  0.434550 -0.234636  \n",
       "CDH2 -2.039772 -0.078417 -0.254197  0.206163 -2.436187 -2.169047 -0.534980  \n",
       "AKT3 -1.726046  0.749765 -2.011875 -1.021904 -2.106272 -0.199360 -0.468198  \n",
       "\n",
       "[5 rows x 17309 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnai_z.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "del rnai_df, rnai_corr, rnai_n, rnai_logp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Old way, based on population distribution of all z-scores. Why is this wrong?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#rnai_mean = rnai_corr.values.mean()\n",
    "#rnai_sd = rnai_corr.values.std()\n",
    "#rnai_z2 = (rnai_corr - rnai_mean) / rnai_sd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CRISPR Data\n",
    "\n",
    "Load CRISPR dependency data and normalize column names to gene names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.1 s, sys: 358 ms, total: 11.5 s\n",
      "Wall time: 11.6 s\n"
     ]
    }
   ],
   "source": [
    "%time crispr_df = pd.read_csv(join(basedir, 'Achilles_gene_effect.csv'), index_col=0)\n",
    "gene_cols = ['%s' % col.split(' ')[0] for col in crispr_df.columns]\n",
    "crispr_df.columns = gene_cols\n",
    "# Drop any duplicate columns (shouldn't be any for CRISPR, but just in case)\n",
    "crispr_df = crispr_df.loc[:,~crispr_df.columns.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000004</th>\n",
       "      <td>0.164686</td>\n",
       "      <td>0.094628</td>\n",
       "      <td>-0.190746</td>\n",
       "      <td>-0.009197</td>\n",
       "      <td>0.032529</td>\n",
       "      <td>-0.191355</td>\n",
       "      <td>0.355279</td>\n",
       "      <td>-0.442430</td>\n",
       "      <td>0.298153</td>\n",
       "      <td>0.170712</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.125219</td>\n",
       "      <td>-0.463670</td>\n",
       "      <td>0.031849</td>\n",
       "      <td>-0.007470</td>\n",
       "      <td>0.261947</td>\n",
       "      <td>0.237661</td>\n",
       "      <td>-0.420546</td>\n",
       "      <td>0.277953</td>\n",
       "      <td>0.227172</td>\n",
       "      <td>-0.151192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000005</th>\n",
       "      <td>-0.097058</td>\n",
       "      <td>0.241606</td>\n",
       "      <td>0.193671</td>\n",
       "      <td>0.156428</td>\n",
       "      <td>-0.195506</td>\n",
       "      <td>-0.337316</td>\n",
       "      <td>0.247947</td>\n",
       "      <td>-0.587806</td>\n",
       "      <td>-0.070959</td>\n",
       "      <td>0.004288</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.212524</td>\n",
       "      <td>-0.410596</td>\n",
       "      <td>-0.180655</td>\n",
       "      <td>-0.080401</td>\n",
       "      <td>0.220940</td>\n",
       "      <td>-0.076012</td>\n",
       "      <td>-0.106125</td>\n",
       "      <td>0.068227</td>\n",
       "      <td>0.022568</td>\n",
       "      <td>-0.261572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000007</th>\n",
       "      <td>0.066212</td>\n",
       "      <td>0.073850</td>\n",
       "      <td>-0.065867</td>\n",
       "      <td>0.156038</td>\n",
       "      <td>0.100958</td>\n",
       "      <td>0.140997</td>\n",
       "      <td>0.066367</td>\n",
       "      <td>-0.472706</td>\n",
       "      <td>-0.008810</td>\n",
       "      <td>0.293417</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.096162</td>\n",
       "      <td>-0.277202</td>\n",
       "      <td>-0.058930</td>\n",
       "      <td>0.108511</td>\n",
       "      <td>0.212320</td>\n",
       "      <td>-0.011173</td>\n",
       "      <td>-0.352653</td>\n",
       "      <td>0.084244</td>\n",
       "      <td>-0.387673</td>\n",
       "      <td>-0.439552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000009</th>\n",
       "      <td>0.097872</td>\n",
       "      <td>0.000361</td>\n",
       "      <td>-0.055818</td>\n",
       "      <td>0.046221</td>\n",
       "      <td>0.063069</td>\n",
       "      <td>-0.048932</td>\n",
       "      <td>0.070910</td>\n",
       "      <td>-0.640566</td>\n",
       "      <td>0.148335</td>\n",
       "      <td>0.056188</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.293175</td>\n",
       "      <td>-0.259388</td>\n",
       "      <td>-0.095524</td>\n",
       "      <td>0.049911</td>\n",
       "      <td>0.072002</td>\n",
       "      <td>0.011488</td>\n",
       "      <td>-0.596006</td>\n",
       "      <td>0.189440</td>\n",
       "      <td>-0.117177</td>\n",
       "      <td>-0.539860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000011</th>\n",
       "      <td>0.270232</td>\n",
       "      <td>0.072105</td>\n",
       "      <td>0.014605</td>\n",
       "      <td>0.433243</td>\n",
       "      <td>-0.030273</td>\n",
       "      <td>-0.239080</td>\n",
       "      <td>0.097866</td>\n",
       "      <td>-0.419689</td>\n",
       "      <td>0.138635</td>\n",
       "      <td>0.106322</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.401794</td>\n",
       "      <td>-0.507831</td>\n",
       "      <td>-0.136369</td>\n",
       "      <td>-0.115541</td>\n",
       "      <td>0.274140</td>\n",
       "      <td>0.154517</td>\n",
       "      <td>-0.234775</td>\n",
       "      <td>0.124355</td>\n",
       "      <td>-0.250505</td>\n",
       "      <td>-0.372498</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18119 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                A1BG      A1CF       A2M     A2ML1   A3GALT2    A4GALT  \\\n",
       "DepMap_ID                                                                \n",
       "ACH-000004  0.164686  0.094628 -0.190746 -0.009197  0.032529 -0.191355   \n",
       "ACH-000005 -0.097058  0.241606  0.193671  0.156428 -0.195506 -0.337316   \n",
       "ACH-000007  0.066212  0.073850 -0.065867  0.156038  0.100958  0.140997   \n",
       "ACH-000009  0.097872  0.000361 -0.055818  0.046221  0.063069 -0.048932   \n",
       "ACH-000011  0.270232  0.072105  0.014605  0.433243 -0.030273 -0.239080   \n",
       "\n",
       "               A4GNT      AAAS      AACS     AADAC  ...    ZWILCH     ZWINT  \\\n",
       "DepMap_ID                                           ...                       \n",
       "ACH-000004  0.355279 -0.442430  0.298153  0.170712  ... -0.125219 -0.463670   \n",
       "ACH-000005  0.247947 -0.587806 -0.070959  0.004288  ... -0.212524 -0.410596   \n",
       "ACH-000007  0.066367 -0.472706 -0.008810  0.293417  ... -0.096162 -0.277202   \n",
       "ACH-000009  0.070910 -0.640566  0.148335  0.056188  ... -0.293175 -0.259388   \n",
       "ACH-000011  0.097866 -0.419689  0.138635  0.106322  ... -0.401794 -0.507831   \n",
       "\n",
       "                ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX  \\\n",
       "DepMap_ID                                                                \n",
       "ACH-000004  0.031849 -0.007470  0.261947  0.237661 -0.420546  0.277953   \n",
       "ACH-000005 -0.180655 -0.080401  0.220940 -0.076012 -0.106125  0.068227   \n",
       "ACH-000007 -0.058930  0.108511  0.212320 -0.011173 -0.352653  0.084244   \n",
       "ACH-000009 -0.095524  0.049911  0.072002  0.011488 -0.596006  0.189440   \n",
       "ACH-000011 -0.136369 -0.115541  0.274140  0.154517 -0.234775  0.124355   \n",
       "\n",
       "               ZZEF1      ZZZ3  \n",
       "DepMap_ID                       \n",
       "ACH-000004  0.227172 -0.151192  \n",
       "ACH-000005  0.022568 -0.261572  \n",
       "ACH-000007 -0.387673 -0.439552  \n",
       "ACH-000009 -0.117177 -0.539860  \n",
       "ACH-000011 -0.250505 -0.372498  \n",
       "\n",
       "[5 rows x 18119 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out this problematic outlier stomach cell line\n",
    "#crispr_df = crispr_df[crispr_df.index != 'ACH-000167']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compute CRISPR correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "recalculate = False\n",
    "    \n",
    "crispr_corr = get_corr(recalculate, crispr_df, 'crispr_correlations')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate CRISPR sample sizes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 sec\n"
     ]
    }
   ],
   "source": [
    "recalculate = False\n",
    "\n",
    "crispr_n = get_n(recalculate, crispr_df, join(basedir, 'crispr_n'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate CRISPR p-values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8.052492141723633 sec\n"
     ]
    }
   ],
   "source": [
    "recalculate = False\n",
    "\n",
    "crispr_logp = get_logp(recalculate, crispr_n, crispr_corr, join(basedir, 'crispr_logp'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>-inf</td>\n",
       "      <td>-2.729089</td>\n",
       "      <td>-3.607518</td>\n",
       "      <td>-1.980992</td>\n",
       "      <td>-0.071521</td>\n",
       "      <td>-2.398956</td>\n",
       "      <td>-0.056143</td>\n",
       "      <td>-0.356653</td>\n",
       "      <td>-0.068348</td>\n",
       "      <td>-2.486415</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.487791</td>\n",
       "      <td>-0.139625</td>\n",
       "      <td>-0.942707</td>\n",
       "      <td>-0.239842</td>\n",
       "      <td>-1.928554</td>\n",
       "      <td>-1.840513</td>\n",
       "      <td>-1.767935</td>\n",
       "      <td>-1.895396</td>\n",
       "      <td>-1.844012</td>\n",
       "      <td>-1.755956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-2.729089</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.698920</td>\n",
       "      <td>-0.432430</td>\n",
       "      <td>-0.002479</td>\n",
       "      <td>-5.172074</td>\n",
       "      <td>-0.001276</td>\n",
       "      <td>-1.934047</td>\n",
       "      <td>-4.393969</td>\n",
       "      <td>-0.775754</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.550875</td>\n",
       "      <td>-0.210882</td>\n",
       "      <td>-4.739521</td>\n",
       "      <td>-3.710275</td>\n",
       "      <td>-1.775766</td>\n",
       "      <td>-0.140062</td>\n",
       "      <td>-1.245317</td>\n",
       "      <td>-3.089883</td>\n",
       "      <td>-0.548347</td>\n",
       "      <td>-1.226263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-3.607518</td>\n",
       "      <td>-0.698920</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-4.959590</td>\n",
       "      <td>-0.624346</td>\n",
       "      <td>-2.082085</td>\n",
       "      <td>-0.231472</td>\n",
       "      <td>-1.223824</td>\n",
       "      <td>-0.122721</td>\n",
       "      <td>-2.952393</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.217154</td>\n",
       "      <td>-2.115227</td>\n",
       "      <td>-0.098963</td>\n",
       "      <td>-2.428453</td>\n",
       "      <td>-0.126596</td>\n",
       "      <td>-0.777269</td>\n",
       "      <td>-0.128080</td>\n",
       "      <td>-0.977968</td>\n",
       "      <td>-2.634127</td>\n",
       "      <td>-1.494651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-1.980992</td>\n",
       "      <td>-0.432430</td>\n",
       "      <td>-4.959590</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-2.025148</td>\n",
       "      <td>-0.460935</td>\n",
       "      <td>-0.555886</td>\n",
       "      <td>-1.550197</td>\n",
       "      <td>-0.033417</td>\n",
       "      <td>-0.563286</td>\n",
       "      <td>...</td>\n",
       "      <td>-8.816429</td>\n",
       "      <td>-4.483147</td>\n",
       "      <td>-4.516258</td>\n",
       "      <td>-3.370472</td>\n",
       "      <td>-0.644246</td>\n",
       "      <td>-0.503514</td>\n",
       "      <td>-0.451465</td>\n",
       "      <td>-0.224883</td>\n",
       "      <td>-1.715529</td>\n",
       "      <td>-0.997601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>-0.071521</td>\n",
       "      <td>-0.002479</td>\n",
       "      <td>-0.624346</td>\n",
       "      <td>-2.025148</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.256902</td>\n",
       "      <td>-1.151857</td>\n",
       "      <td>-2.246387</td>\n",
       "      <td>-0.305064</td>\n",
       "      <td>-0.162205</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.027283</td>\n",
       "      <td>-0.496897</td>\n",
       "      <td>-0.532855</td>\n",
       "      <td>-0.532775</td>\n",
       "      <td>-11.028198</td>\n",
       "      <td>-1.635245</td>\n",
       "      <td>-1.599105</td>\n",
       "      <td>-0.067629</td>\n",
       "      <td>-0.483409</td>\n",
       "      <td>-2.461702</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18119 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             A1BG      A1CF       A2M     A2ML1   A3GALT2    A4GALT     A4GNT  \\\n",
       "A1BG         -inf -2.729089 -3.607518 -1.980992 -0.071521 -2.398956 -0.056143   \n",
       "A1CF    -2.729089      -inf -0.698920 -0.432430 -0.002479 -5.172074 -0.001276   \n",
       "A2M     -3.607518 -0.698920      -inf -4.959590 -0.624346 -2.082085 -0.231472   \n",
       "A2ML1   -1.980992 -0.432430 -4.959590      -inf -2.025148 -0.460935 -0.555886   \n",
       "A3GALT2 -0.071521 -0.002479 -0.624346 -2.025148      -inf -0.256902 -1.151857   \n",
       "\n",
       "             AAAS      AACS     AADAC  ...    ZWILCH     ZWINT      ZXDA  \\\n",
       "A1BG    -0.356653 -0.068348 -2.486415  ... -0.487791 -0.139625 -0.942707   \n",
       "A1CF    -1.934047 -4.393969 -0.775754  ... -4.550875 -0.210882 -4.739521   \n",
       "A2M     -1.223824 -0.122721 -2.952393  ... -0.217154 -2.115227 -0.098963   \n",
       "A2ML1   -1.550197 -0.033417 -0.563286  ... -8.816429 -4.483147 -4.516258   \n",
       "A3GALT2 -2.246387 -0.305064 -0.162205  ... -1.027283 -0.496897 -0.532855   \n",
       "\n",
       "             ZXDB       ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
       "A1BG    -0.239842  -1.928554 -1.840513 -1.767935 -1.895396 -1.844012 -1.755956  \n",
       "A1CF    -3.710275  -1.775766 -0.140062 -1.245317 -3.089883 -0.548347 -1.226263  \n",
       "A2M     -2.428453  -0.126596 -0.777269 -0.128080 -0.977968 -2.634127 -1.494651  \n",
       "A2ML1   -3.370472  -0.644246 -0.503514 -0.451465 -0.224883 -1.715529 -0.997601  \n",
       "A3GALT2 -0.532775 -11.028198 -1.635245 -1.599105 -0.067629 -0.483409 -2.461702  \n",
       "\n",
       "[5 rows x 18119 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_logp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mplayer\n",
    "p = mplayer.Player()\n",
    "success_file = '/Users/johnbachman/Downloads/success.m4a'\n",
    "p.loadfile(success_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate CRISPR z-scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8.676475048065186 sec\n"
     ]
    }
   ],
   "source": [
    "recalculate = False\n",
    "\n",
    "#crispr_z_ppf = get_z(recalculate, crispr_logp, crispr_corr, 'crispr_z')\n",
    "crispr_z = get_z(recalculate, crispr_logp, crispr_corr, join(basedir, 'crispr_z_log'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Old way, using population distribution of correlation coefficients:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#crispr_mean = crispr_corr.values.mean()\n",
    "#crispr_sd = crispr_corr.values.std()\n",
    "#crispr_z = (crispr_corr - crispr_mean) / crispr_sd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clear memory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "del crispr_df, crispr_corr, crispr_n, crispr_logp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### Combine CRISPR and RNAi Z-scores\n",
    " \n",
    " Stouffer's formula:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "dep_z = (crispr_z + rnai_z) / np.sqrt(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many infinities due to underflow?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15669"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(dep_z.abs() == np.inf).sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many NaN values are there in the combined dataset?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "145042232"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.isna(dep_z).sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>inf</td>\n",
       "      <td>-2.344420</td>\n",
       "      <td>1.215558</td>\n",
       "      <td>-1.562173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.092157</td>\n",
       "      <td>0.348988</td>\n",
       "      <td>0.956904</td>\n",
       "      <td>-1.551909</td>\n",
       "      <td>1.723844</td>\n",
       "      <td>...</td>\n",
       "      <td>2.461753</td>\n",
       "      <td>-0.997734</td>\n",
       "      <td>-0.392137</td>\n",
       "      <td>-0.206969</td>\n",
       "      <td>-1.259928</td>\n",
       "      <td>1.499647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.794664</td>\n",
       "      <td>0.817246</td>\n",
       "      <td>0.481082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-2.344420</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.176680</td>\n",
       "      <td>1.357196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.648383</td>\n",
       "      <td>0.721116</td>\n",
       "      <td>2.584396</td>\n",
       "      <td>-0.877176</td>\n",
       "      <td>2.035648</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.393022</td>\n",
       "      <td>-0.746085</td>\n",
       "      <td>1.304967</td>\n",
       "      <td>2.017047</td>\n",
       "      <td>-0.127186</td>\n",
       "      <td>2.306467</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.442976</td>\n",
       "      <td>-1.416195</td>\n",
       "      <td>-0.541671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>1.215558</td>\n",
       "      <td>0.176680</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.005903</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.946165</td>\n",
       "      <td>1.592105</td>\n",
       "      <td>-1.093477</td>\n",
       "      <td>-1.687519</td>\n",
       "      <td>-0.266744</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.530944</td>\n",
       "      <td>-0.450800</td>\n",
       "      <td>-1.249026</td>\n",
       "      <td>-0.383748</td>\n",
       "      <td>-1.994584</td>\n",
       "      <td>-0.470479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.543282</td>\n",
       "      <td>1.312836</td>\n",
       "      <td>-0.777776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-1.562173</td>\n",
       "      <td>1.357196</td>\n",
       "      <td>0.005903</td>\n",
       "      <td>inf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.337678</td>\n",
       "      <td>-0.333557</td>\n",
       "      <td>0.363633</td>\n",
       "      <td>1.377973</td>\n",
       "      <td>1.380592</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.583980</td>\n",
       "      <td>-1.600073</td>\n",
       "      <td>-0.268438</td>\n",
       "      <td>-2.212961</td>\n",
       "      <td>0.705614</td>\n",
       "      <td>0.269420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.010898</td>\n",
       "      <td>-1.821388</td>\n",
       "      <td>-1.744558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19761 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             A1BG      A1CF       A2M     A2ML1  A3GALT2    A4GALT     A4GNT  \\\n",
       "A1BG          inf -2.344420  1.215558 -1.562173      NaN  1.092157  0.348988   \n",
       "A1CF    -2.344420       inf  0.176680  1.357196      NaN -2.648383  0.721116   \n",
       "A2M      1.215558  0.176680       inf  0.005903      NaN -1.946165  1.592105   \n",
       "A2ML1   -1.562173  1.357196  0.005903       inf      NaN  0.337678 -0.333557   \n",
       "A3GALT2       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
       "\n",
       "             AAAS      AACS     AADAC  ...    ZWILCH     ZWINT      ZXDA  \\\n",
       "A1BG     0.956904 -1.551909  1.723844  ...  2.461753 -0.997734 -0.392137   \n",
       "A1CF     2.584396 -0.877176  2.035648  ... -3.393022 -0.746085  1.304967   \n",
       "A2M     -1.093477 -1.687519 -0.266744  ... -1.530944 -0.450800 -1.249026   \n",
       "A2ML1    0.363633  1.377973  1.380592  ... -2.583980 -1.600073 -0.268438   \n",
       "A3GALT2       NaN       NaN       NaN  ...       NaN       NaN       NaN   \n",
       "\n",
       "             ZXDB      ZXDC    ZYG11A  ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
       "A1BG    -0.206969 -1.259928  1.499647     NaN  0.794664  0.817246  0.481082  \n",
       "A1CF     2.017047 -0.127186  2.306467     NaN  0.442976 -1.416195 -0.541671  \n",
       "A2M     -0.383748 -1.994584 -0.470479     NaN -0.543282  1.312836 -0.777776  \n",
       "A2ML1   -2.212961  0.705614  0.269420     NaN -1.010898 -1.821388 -1.744558  \n",
       "A3GALT2       NaN       NaN       NaN     NaN       NaN       NaN       NaN  \n",
       "\n",
       "[5 rows x 19761 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dep_z.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The large number of NaN values are due to columns being present in RNAi or CRISPR but not both. We drop the NaN rows/columns here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dep_z = dep_z.dropna(axis=0, how='all').dropna(axis=1, how='all')\n",
    "pd.isna(dep_z).sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the z-scores back into p-values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "dep_logp = pd.DataFrame(np.log(2) + stats.norm.logcdf(-dep_z.abs()),\n",
    "                        index=dep_z.columns, columns=dep_z.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>-inf</td>\n",
       "      <td>-3.960337</td>\n",
       "      <td>-1.495426</td>\n",
       "      <td>-2.134978</td>\n",
       "      <td>-1.291842</td>\n",
       "      <td>-0.318694</td>\n",
       "      <td>-1.082890</td>\n",
       "      <td>-2.114580</td>\n",
       "      <td>-2.468216</td>\n",
       "      <td>-2.178407</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.797799</td>\n",
       "      <td>-4.281204</td>\n",
       "      <td>-1.144420</td>\n",
       "      <td>-0.363905</td>\n",
       "      <td>-0.179086</td>\n",
       "      <td>-1.571683</td>\n",
       "      <td>-2.012112</td>\n",
       "      <td>-0.851419</td>\n",
       "      <td>-0.882402</td>\n",
       "      <td>-0.461308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-3.960337</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.151102</td>\n",
       "      <td>-1.744577</td>\n",
       "      <td>-4.817401</td>\n",
       "      <td>-0.753241</td>\n",
       "      <td>-4.629978</td>\n",
       "      <td>-0.966556</td>\n",
       "      <td>-3.175200</td>\n",
       "      <td>-0.083723</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.720491</td>\n",
       "      <td>-7.276993</td>\n",
       "      <td>-0.786105</td>\n",
       "      <td>-1.650760</td>\n",
       "      <td>-3.130621</td>\n",
       "      <td>-0.106702</td>\n",
       "      <td>-3.859215</td>\n",
       "      <td>-0.418880</td>\n",
       "      <td>-1.853304</td>\n",
       "      <td>-0.530951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-1.495426</td>\n",
       "      <td>-0.151102</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.004721</td>\n",
       "      <td>-2.963557</td>\n",
       "      <td>-2.194976</td>\n",
       "      <td>-1.293954</td>\n",
       "      <td>-2.391376</td>\n",
       "      <td>-0.236145</td>\n",
       "      <td>-2.723761</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.032754</td>\n",
       "      <td>-2.073195</td>\n",
       "      <td>-0.427505</td>\n",
       "      <td>-1.552796</td>\n",
       "      <td>-0.355012</td>\n",
       "      <td>-3.077196</td>\n",
       "      <td>-0.449397</td>\n",
       "      <td>-0.532840</td>\n",
       "      <td>-1.664749</td>\n",
       "      <td>-0.828507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-2.134978</td>\n",
       "      <td>-1.744577</td>\n",
       "      <td>-0.004721</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.307061</td>\n",
       "      <td>-0.302845</td>\n",
       "      <td>-0.333891</td>\n",
       "      <td>-1.782532</td>\n",
       "      <td>-1.787342</td>\n",
       "      <td>-0.086286</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.348286</td>\n",
       "      <td>-4.628771</td>\n",
       "      <td>-2.211079</td>\n",
       "      <td>-0.237797</td>\n",
       "      <td>-3.615617</td>\n",
       "      <td>-0.733077</td>\n",
       "      <td>-0.238757</td>\n",
       "      <td>-1.164542</td>\n",
       "      <td>-2.680222</td>\n",
       "      <td>-2.512542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>-1.291842</td>\n",
       "      <td>-4.817401</td>\n",
       "      <td>-2.963557</td>\n",
       "      <td>-0.307061</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.066640</td>\n",
       "      <td>-3.306363</td>\n",
       "      <td>-0.627601</td>\n",
       "      <td>-0.913968</td>\n",
       "      <td>-4.724983</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.128864</td>\n",
       "      <td>-0.179719</td>\n",
       "      <td>-0.505921</td>\n",
       "      <td>-0.366647</td>\n",
       "      <td>-1.474962</td>\n",
       "      <td>-1.692142</td>\n",
       "      <td>-1.132102</td>\n",
       "      <td>-1.183917</td>\n",
       "      <td>-2.753599</td>\n",
       "      <td>-0.492305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>-1.571683</td>\n",
       "      <td>-0.106702</td>\n",
       "      <td>-3.077196</td>\n",
       "      <td>-0.733077</td>\n",
       "      <td>-1.692142</td>\n",
       "      <td>-0.430116</td>\n",
       "      <td>-3.692502</td>\n",
       "      <td>-2.104900</td>\n",
       "      <td>-0.053739</td>\n",
       "      <td>-2.269848</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.839801</td>\n",
       "      <td>-0.529897</td>\n",
       "      <td>-6.057934</td>\n",
       "      <td>-1.179455</td>\n",
       "      <td>-0.637011</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-2.175763</td>\n",
       "      <td>-0.234696</td>\n",
       "      <td>-2.445311</td>\n",
       "      <td>-2.157461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>-2.012112</td>\n",
       "      <td>-3.859215</td>\n",
       "      <td>-0.449397</td>\n",
       "      <td>-0.238757</td>\n",
       "      <td>-1.132102</td>\n",
       "      <td>-3.445092</td>\n",
       "      <td>-0.096436</td>\n",
       "      <td>-3.551059</td>\n",
       "      <td>-0.690674</td>\n",
       "      <td>-1.295834</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.902272</td>\n",
       "      <td>-1.502313</td>\n",
       "      <td>-1.426861</td>\n",
       "      <td>-1.817710</td>\n",
       "      <td>-1.882839</td>\n",
       "      <td>-2.175763</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.016753</td>\n",
       "      <td>-3.809342</td>\n",
       "      <td>-0.097269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>-0.851419</td>\n",
       "      <td>-0.418880</td>\n",
       "      <td>-0.532840</td>\n",
       "      <td>-1.164542</td>\n",
       "      <td>-1.183917</td>\n",
       "      <td>-0.483968</td>\n",
       "      <td>-2.431931</td>\n",
       "      <td>-4.389975</td>\n",
       "      <td>-2.138405</td>\n",
       "      <td>-16.669678</td>\n",
       "      <td>...</td>\n",
       "      <td>-10.844030</td>\n",
       "      <td>-7.859672</td>\n",
       "      <td>-1.039342</td>\n",
       "      <td>-0.171372</td>\n",
       "      <td>-7.555381</td>\n",
       "      <td>-0.234696</td>\n",
       "      <td>-0.016753</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-4.705826</td>\n",
       "      <td>-0.999728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>-0.882402</td>\n",
       "      <td>-1.853304</td>\n",
       "      <td>-1.664749</td>\n",
       "      <td>-2.680222</td>\n",
       "      <td>-2.753599</td>\n",
       "      <td>-1.824835</td>\n",
       "      <td>-1.367785</td>\n",
       "      <td>-1.943995</td>\n",
       "      <td>-0.077415</td>\n",
       "      <td>-3.326127</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.688584</td>\n",
       "      <td>-1.011499</td>\n",
       "      <td>-4.596264</td>\n",
       "      <td>-0.400738</td>\n",
       "      <td>-0.427752</td>\n",
       "      <td>-2.445311</td>\n",
       "      <td>-3.809342</td>\n",
       "      <td>-4.705826</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-2.700534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>-0.461308</td>\n",
       "      <td>-0.530951</td>\n",
       "      <td>-0.828507</td>\n",
       "      <td>-2.512542</td>\n",
       "      <td>-0.492305</td>\n",
       "      <td>-0.871511</td>\n",
       "      <td>-1.438407</td>\n",
       "      <td>-2.549401</td>\n",
       "      <td>-1.810054</td>\n",
       "      <td>-0.940322</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.687093</td>\n",
       "      <td>-2.379520</td>\n",
       "      <td>-1.192396</td>\n",
       "      <td>-0.657296</td>\n",
       "      <td>-0.565062</td>\n",
       "      <td>-2.157461</td>\n",
       "      <td>-0.097269</td>\n",
       "      <td>-0.999728</td>\n",
       "      <td>-2.700534</td>\n",
       "      <td>-inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15667 rows × 15667 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            A1BG      A1CF       A2M     A2ML1    A4GALT     A4GNT      AAAS  \\\n",
       "A1BG        -inf -3.960337 -1.495426 -2.134978 -1.291842 -0.318694 -1.082890   \n",
       "A1CF   -3.960337      -inf -0.151102 -1.744577 -4.817401 -0.753241 -4.629978   \n",
       "A2M    -1.495426 -0.151102      -inf -0.004721 -2.963557 -2.194976 -1.293954   \n",
       "A2ML1  -2.134978 -1.744577 -0.004721      -inf -0.307061 -0.302845 -0.333891   \n",
       "A4GALT -1.291842 -4.817401 -2.963557 -0.307061      -inf -0.066640 -3.306363   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "ZXDC   -1.571683 -0.106702 -3.077196 -0.733077 -1.692142 -0.430116 -3.692502   \n",
       "ZYG11A -2.012112 -3.859215 -0.449397 -0.238757 -1.132102 -3.445092 -0.096436   \n",
       "ZYX    -0.851419 -0.418880 -0.532840 -1.164542 -1.183917 -0.483968 -2.431931   \n",
       "ZZEF1  -0.882402 -1.853304 -1.664749 -2.680222 -2.753599 -1.824835 -1.367785   \n",
       "ZZZ3   -0.461308 -0.530951 -0.828507 -2.512542 -0.492305 -0.871511 -1.438407   \n",
       "\n",
       "            AACS     AADAC    AADACL2  ...       ZW10    ZWILCH     ZWINT  \\\n",
       "A1BG   -2.114580 -2.468216  -2.178407  ...  -1.797799 -4.281204 -1.144420   \n",
       "A1CF   -0.966556 -3.175200  -0.083723  ...  -0.720491 -7.276993 -0.786105   \n",
       "A2M    -2.391376 -0.236145  -2.723761  ...  -0.032754 -2.073195 -0.427505   \n",
       "A2ML1  -1.782532 -1.787342  -0.086286  ...  -1.348286 -4.628771 -2.211079   \n",
       "A4GALT -0.627601 -0.913968  -4.724983  ...  -1.128864 -0.179719 -0.505921   \n",
       "...          ...       ...        ...  ...        ...       ...       ...   \n",
       "ZXDC   -2.104900 -0.053739  -2.269848  ...  -2.839801 -0.529897 -6.057934   \n",
       "ZYG11A -3.551059 -0.690674  -1.295834  ...  -0.902272 -1.502313 -1.426861   \n",
       "ZYX    -4.389975 -2.138405 -16.669678  ... -10.844030 -7.859672 -1.039342   \n",
       "ZZEF1  -1.943995 -0.077415  -3.326127  ...  -0.688584 -1.011499 -4.596264   \n",
       "ZZZ3   -2.549401 -1.810054  -0.940322  ...  -1.687093 -2.379520 -1.192396   \n",
       "\n",
       "            ZXDA      ZXDB      ZXDC    ZYG11A       ZYX     ZZEF1      ZZZ3  \n",
       "A1BG   -0.363905 -0.179086 -1.571683 -2.012112 -0.851419 -0.882402 -0.461308  \n",
       "A1CF   -1.650760 -3.130621 -0.106702 -3.859215 -0.418880 -1.853304 -0.530951  \n",
       "A2M    -1.552796 -0.355012 -3.077196 -0.449397 -0.532840 -1.664749 -0.828507  \n",
       "A2ML1  -0.237797 -3.615617 -0.733077 -0.238757 -1.164542 -2.680222 -2.512542  \n",
       "A4GALT -0.366647 -1.474962 -1.692142 -1.132102 -1.183917 -2.753599 -0.492305  \n",
       "...          ...       ...       ...       ...       ...       ...       ...  \n",
       "ZXDC   -1.179455 -0.637011      -inf -2.175763 -0.234696 -2.445311 -2.157461  \n",
       "ZYG11A -1.817710 -1.882839 -2.175763      -inf -0.016753 -3.809342 -0.097269  \n",
       "ZYX    -0.171372 -7.555381 -0.234696 -0.016753      -inf -4.705826 -0.999728  \n",
       "ZZEF1  -0.400738 -0.427752 -2.445311 -3.809342 -4.705826      -inf -2.700534  \n",
       "ZZZ3   -0.657296 -0.565062 -2.157461 -0.097269 -0.999728 -2.700534      -inf  \n",
       "\n",
       "[15667 rows x 15667 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dep_logp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the z-scores:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "recalculate = False\n",
    "filename = 'dep_z'\n",
    "z_filepath = join(basedir, '%s.h5' % filename)\n",
    "if recalculate:\n",
    "    dep_z.to_hdf(z_filepath, filename)\n",
    "else:\n",
    "    dep_z = pd.read_hdf(z_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Old way: take the average of CRISPR and RNAi z-scores:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "#recalculate = False\n",
    "#filename = 'dep_z'\n",
    "#z_filepath = join(basedir, '%s.h5' % filename)\n",
    "#if recalculate:\n",
    "#    %time dep_z = (crispr_z + rnai_z) / 2\n",
    "#    #dep_z = dep_z.dropna(axis=0, how='all').dropna(axis=1, how='all')\n",
    "#    dep_z.to_hdf(z_filepath, filename)\n",
    "#else:\n",
    "#    dep_z = pd.read_hdf(z_filepath)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the p-values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "recalculate = True\n",
    "filename = 'dep_logp'\n",
    "logp_filepath = join(basedir, '%s.h5' % filename)\n",
    "if recalculate:\n",
    "    dep_logp.to_hdf(logp_filepath, filename)\n",
    "else:\n",
    "    dep_logp = pd.read_hdf(z_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get correlations with correction values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A useful function for looking at filtered sets of correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unstack_corrs(df):\n",
    "    df_ut = df.where(np.triu(np.ones(df.shape), k=1).astype(np.bool))\n",
    "    stacked: pd.DataFrame = df_ut.stack(dropna=True)\n",
    "    return stacked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify which dataset of p-values we'll be testing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_logp = crispr_logp\n",
    "df_logp = dep_logp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the number of gene pairs that we'll be testing. First we get the total number of correlations in the upper triangular part of the matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "122719611"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_comps = np.triu(~df_logp.isna(), k=1).sum()\n",
    "total_comps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we calculate how many correlations are between strictly mitochondrial genes. Because we won't be considering these correlations, we subtract these from our number of comparisons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "121778711"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mitogenes_in_df = set(df_logp.columns).intersection(set(mitogenes))\n",
    "mito_comps = len(mitogenes_in_df)**2\n",
    "num_comps = total_comps - mito_comps\n",
    "num_comps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Calculate the log p-value threshold of alpha / num_comps:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-21.61344838498181"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alpha = 0.05\n",
    "bc_thresh = np.log(alpha / num_comps)\n",
    "bc_thresh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get all correlations with uncorrected p-values below threshold:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_no_corr = unstack_corrs(df_logp[df_logp < np.log(alpha)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A1BG    A1CF    -3.960337\n",
       "        AAMP    -4.451078\n",
       "        AANAT   -5.925051\n",
       "        AARS2   -9.828733\n",
       "        ABCA1   -3.829890\n",
       "                   ...   \n",
       "ZWINT   ZZEF1   -4.596264\n",
       "ZXDA    ZXDB    -9.143097\n",
       "ZXDB    ZYX     -7.555381\n",
       "ZYG11A  ZZEF1   -3.809342\n",
       "ZYX     ZZEF1   -4.705826\n",
       "Length: 25343861, dtype: float64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_no_corr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out correlations between mitogenes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filt_mitocorrs(ser):\n",
    "    filt_ix = []\n",
    "    filt_vals = []\n",
    "    for (ix0, ix1), logp in ser.iteritems():\n",
    "        if ix0 in mitogenes and ix1 in mitogenes:\n",
    "            continue\n",
    "        filt_ix.append((ix0, ix1))\n",
    "        filt_vals.append(logp)\n",
    "    index = pd.MultiIndex.from_tuples(filt_ix, names=['geneA', 'geneB'])\n",
    "    filt_ser = pd.Series(filt_vals, index=index)\n",
    "    return filt_ser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_corrs = filt_mitocorrs(sig_no_corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "geneA   geneB\n",
       "A1BG    A1CF    -3.960337\n",
       "        AAMP    -4.451078\n",
       "        AANAT   -5.925051\n",
       "        AARS2   -9.828733\n",
       "        ABCA1   -3.829890\n",
       "                   ...   \n",
       "ZWINT   ZZEF1   -4.596264\n",
       "ZXDA    ZXDB    -9.143097\n",
       "ZXDB    ZYX     -7.555381\n",
       "ZYG11A  ZZEF1   -3.809342\n",
       "ZYX     ZZEF1   -4.705826\n",
       "Length: 25211483, dtype: float64"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt_corrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mplayer\n",
    "p = mplayer.Player()\n",
    "success_file = '/Users/johnbachman/Downloads/success.m4a'\n",
    "p.loadfile(success_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Pairs at corrected threshold for CRISPR: 168,098\n",
    "* For RNAi: 142,430\n",
    "* For combined by Stouffer: ~62,320"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate correction thresholds (Bonferroni, B-H, B-Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_sorted = filt_corrs.sort_values().to_frame('logp')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add rank column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_sorted['rank'] = sig_sorted.rank()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bonferroni correction threshold:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_sorted['bc_cutoff'] = bc_thresh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Benjamini-Hochberg critical value:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_sorted['bh_crit_val'] = np.log((sig_sorted['rank'] / num_comps) * alpha)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Benjamini-Yekutieli critical value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "cm = np.log(num_comps) + np.euler_gamma + (1/(2*num_comps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_sorted['by_crit_val'] = sig_sorted['bh_crit_val'] - np.log(cm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>logp</th>\n",
       "      <th>rank</th>\n",
       "      <th>bc_cutoff</th>\n",
       "      <th>bh_crit_val</th>\n",
       "      <th>by_crit_val</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneA</th>\n",
       "      <th>geneB</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSC1</th>\n",
       "      <th>TSC2</th>\n",
       "      <td>-inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-24.568095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MDM2</th>\n",
       "      <th>TP53</th>\n",
       "      <td>-585.802559</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-20.920301</td>\n",
       "      <td>-23.874947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <th>TP53BP1</th>\n",
       "      <td>-511.445963</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-20.514836</td>\n",
       "      <td>-23.469482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP2M1</th>\n",
       "      <th>AP2S1</th>\n",
       "      <td>-464.574498</td>\n",
       "      <td>4.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-20.227154</td>\n",
       "      <td>-23.181800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDKN1A</th>\n",
       "      <th>TP53</th>\n",
       "      <td>-444.180749</td>\n",
       "      <td>5.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-20.004010</td>\n",
       "      <td>-22.958657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAPT</th>\n",
       "      <th>MAST1</th>\n",
       "      <td>-2.995733</td>\n",
       "      <td>25211479.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-4.570638</td>\n",
       "      <td>-7.525285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC58</th>\n",
       "      <th>PNMA6A</th>\n",
       "      <td>-2.995733</td>\n",
       "      <td>25211480.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-4.570638</td>\n",
       "      <td>-7.525285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPRN</th>\n",
       "      <th>ZBTB2</th>\n",
       "      <td>-2.995733</td>\n",
       "      <td>25211481.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-4.570638</td>\n",
       "      <td>-7.525285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPLX2</th>\n",
       "      <th>TMEM171</th>\n",
       "      <td>-2.995733</td>\n",
       "      <td>25211482.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-4.570638</td>\n",
       "      <td>-7.525285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPP1R1B</th>\n",
       "      <th>RTP5</th>\n",
       "      <td>-2.995732</td>\n",
       "      <td>25211483.0</td>\n",
       "      <td>-21.613448</td>\n",
       "      <td>-4.570638</td>\n",
       "      <td>-7.525285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25211483 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       logp        rank  bc_cutoff  bh_crit_val  by_crit_val\n",
       "geneA   geneB                                                               \n",
       "TSC1    TSC2           -inf         1.0 -21.613448   -21.613448   -24.568095\n",
       "MDM2    TP53    -585.802559         2.0 -21.613448   -20.920301   -23.874947\n",
       "TP53    TP53BP1 -511.445963         3.0 -21.613448   -20.514836   -23.469482\n",
       "AP2M1   AP2S1   -464.574498         4.0 -21.613448   -20.227154   -23.181800\n",
       "CDKN1A  TP53    -444.180749         5.0 -21.613448   -20.004010   -22.958657\n",
       "...                     ...         ...        ...          ...          ...\n",
       "GAPT    MAST1     -2.995733  25211479.0 -21.613448    -4.570638    -7.525285\n",
       "CCDC58  PNMA6A    -2.995733  25211480.0 -21.613448    -4.570638    -7.525285\n",
       "PTPRN   ZBTB2     -2.995733  25211481.0 -21.613448    -4.570638    -7.525285\n",
       "CPLX2   TMEM171   -2.995733  25211482.0 -21.613448    -4.570638    -7.525285\n",
       "PPP1R1B RTP5      -2.995732  25211483.0 -21.613448    -4.570638    -7.525285\n",
       "\n",
       "[25211483 rows x 5 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Correlations sig by Bonferroni:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(86948, 5)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_sorted[sig_sorted['logp'] < sig_sorted['bc_cutoff']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7349009, 5)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_sorted[sig_sorted['logp'] < sig_sorted['bh_crit_val']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1279296, 5)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_sorted[sig_sorted['logp'] < sig_sorted['by_crit_val']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = join(basedir, 'dep_stouffer_signif.pkl')\n",
    "sig_sorted.to_pickle(filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### Expression Data\n",
    "\n",
    "Load the RNA-seq data (in transcripts per million, obtained from the DepMap website at https://depmap.org/portal/download/). Normalize column names (which include both gene names and Ensemble IDs) to only gene names, and then drop duplicate columns, keeping the first instance of the column for the gene name (there are 87 genes associated with more than one Ensemble ID, and these result in duplicate columns after normalizing to gene names)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%time ccle_df = pd.read_csv(join(basedir, 'CCLE_expression.csv'), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "gene_cols = ['%s' % col.split(' ')[0] for col in ccle_df.columns]\n",
    "ccle_df.columns = gene_cols\n",
    "# NOTE: This first way doesn't work! Probably because with duplicate column names a\n",
    "# separate integer index is created after transposition (?)\n",
    "#%time ccle_df_drop = ccle_df.T.drop_duplicates(keep='first').T\n",
    "# See: https://stackoverflow.com/questions/14984119/python-pandas-remove-duplicate-columns\n",
    "ccle_df_drop = ccle_df.loc[:,~ccle_df.columns.duplicated()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Calculate expression correlation matrix (takes about 11 min):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "recalculate = False\n",
    "filename = 'expr_correlations'\n",
    "filepath = join(basedir, filename)\n",
    "if recalculate:\n",
    "    %time expr_corr = ccle_df.corr()\n",
    "    expr_corr.to_hdf('%s.h5' % filepath, filename)\n",
    "else:\n",
    "    expr_corr = pd.read_hdf('%s.h5' % filepath)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Filter the expression data to contain only those columns also contained in the CRISPR dependency data (protein coding genes, for the most part)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "crispr_cols = [c for c in crispr_df.columns]\n",
    "# Create a boolean mask to filter the CCLE expression data columns\n",
    "ccle_col_mask = np.array(ccle_df_drop.columns.map(lambda x: x in crispr_cols), dtype=bool)\n",
    "ccle_df_filt = ccle_df_drop[ccle_df_drop.columns[ccle_col_mask]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Join the two datasets on the cell line index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "crispr_expr_join_df = crispr_df.join(ccle_df_filt, how='left', lsuffix=' KO', rsuffix=' RNA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Calculate the correlations (slow) or reload from HDF5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "recalculate = False\n",
    "filename = 'crispr_expr_corr'\n",
    "filepath = join(basedir, filename)\n",
    "if recalculate:\n",
    "    %time ce_corr = crispr_expr_join_df.corr()\n",
    "    ce_corr.to_hdf('%s.h5' % filepath, filename)\n",
    "else:\n",
    "    ce_corr = pd.read_hdf('%s.h5' % filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Drug Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_resp_df = pd.read_csv(join(basedir, 'primary-screen-replicate-collapsed-logfold-change.csv'), index_col=0)\n",
    "drug_info_df = pd.read_csv(join(basedir, 'primary-screen-replicate-collapsed-treatment-info.csv'), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_resp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Join drug sensitivity to expression on the cell line index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_expr_df = drug_resp_df.join(ccle_df_filt, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "drug_expr_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define functions for exploring and visualizing the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the sorted list of correlations for a given gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def sort_corrs(corrs, col_name):\n",
    "    s = corrs[col_name]\n",
    "    return sorted(list(zip(s.index, s)), key=lambda x: abs(float(x[1])), reverse=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a function to get the top correlated genes for a given gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def get_expr_corrs(df, geneX):\n",
    "    mygene_arr = df[geneX].values\n",
    "    data = df.values\n",
    "    corrs = np.zeros(data.shape[1])\n",
    "    for i in range(data.shape[1]):\n",
    "        vect = data[:,i]\n",
    "        r = np.corrcoef(mygene_arr, vect)[0, 1]\n",
    "        corrs[i] = r\n",
    "    genes = [c for c in df.columns]\n",
    "    corr_genes = sorted(list(zip(genes, corrs)), reverse=True, key=lambda x: abs(x[1]))\n",
    "    corr_genes_dict = {k: v for k, v in corr_genes}\n",
    "    return corr_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def corrs_for_genes(genes, df=None, cutoff=3.0):\n",
    "    all_corrs = []\n",
    "    if df is None:\n",
    "        df = dep_z\n",
    "    for gene_a in genes:\n",
    "        try:\n",
    "            gene_corrs = sort_corrs(df, gene_a)[1:]\n",
    "        except KeyError:\n",
    "            continue\n",
    "        for gene_b, corr_z in gene_corrs:\n",
    "            if abs(corr_z > cutoff):\n",
    "                all_corrs.append((gene_a, gene_b, corr_z))\n",
    "    return all_corrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corrs_for_genes(['SLC43A1', 'SLC43A2'], cutoff=1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sort_corrs(expr_corr, 'LATS1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a function to plot the relationship between the transcription of two genes, along with a linear regression and the residuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def scatter(df, geneX, y_arr, y_label, bins=30):\n",
    "    plt.figure(figsize=(12,3))\n",
    "    plt.subplot(1, 3, 1)\n",
    "    plt.plot(df[geneX], y_arr, linestyle='', marker='.')\n",
    "    plt.xlabel('%s TPM' % geneX)\n",
    "    plt.ylabel(y_label)\n",
    "    r = np.corrcoef(df[geneX], y_arr)[0][1]\n",
    "    plt.title('Corr: %.3f' % r)\n",
    "    slope, intercept, r_value, p_value, std_err = \\\n",
    "                    stats.linregress(df[geneX], y_arr)\n",
    "    yhat = slope * df[geneX] + intercept\n",
    "    plt.plot(df[geneX], yhat, color='k')\n",
    "    residuals = y_arr - yhat\n",
    "    plt.subplot(1, 3, 2)\n",
    "    plt.hist(residuals, bins=bins)\n",
    "    plt.title('Residuals')\n",
    "    plt.subplot(1, 3, 3)\n",
    "    stats.probplot(residuals, dist='norm', plot=plt)\n",
    "    plt.title('Quantile-quantile plot vs. normal')\n",
    "    return residuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res1 = scatter(ccle_df, 'AKT1', ccle_df['BAX'], 'AKT1 TPM', bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res2 = scatter(ccle_df, 'YAP1', res1, 'Res 1', bins=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next steps:\n",
    "1. load the INDRA SIF statement file\n",
    "2. tabulate explanations for correlations;\n",
    "3. predict correlations from graph structure, edge properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sorted(corrs_for_genes(['KIRREL1'], df=expr_corr), \n",
    "       key=lambda x: x[2], reverse=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paper Figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "opath = '/Users/johnbachman/Dropbox/1johndata/Knowledge File/Biology/Research/Big Mechanism' \\\n",
    "        '/bioexp_paper/output'\n",
    "prefix = 'fig6_ipynb'\n",
    "\n",
    "def fig_path(name, fmt):\n",
    "    return join(opath, f'{prefix}_{name}.{fmt}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure: Example correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_example(geneA, geneB):\n",
    "    lw = 0.5\n",
    "    ms = 1\n",
    "    fig = plt.figure(figsize=(4, 2), dpi=150)\n",
    "    \n",
    "    # CRISPR plot\n",
    "    plt.subplot(1, 2, 1)\n",
    "    ax = plt.gca()\n",
    "    # Axes lines\n",
    "    ax.axhline(y=0, color='k', linewidth=lw)\n",
    "    ax.axvline(x=0, color='k', linewidth=lw)\n",
    "    plt.plot(crispr_df[geneA].values, crispr_df[geneB].values, marker='.', markersize=ms, linestyle='')\n",
    "    \n",
    "    # Plot linear regression\n",
    "    crispr_lr = linregress(crispr_df[geneA].values, crispr_df[geneB].values)\n",
    "    plt.plot(crispr_df[geneA].values, crispr_df[geneA].values * crispr_lr.slope + crispr_lr.intercept,\n",
    "             linestyle='-', linewidth=lw, color='black')\n",
    "    plt.xlabel(f'{geneA} Gene Effect (CERES)')\n",
    "    plt.ylabel(f'{geneB} Gene Effect (CERES)')\n",
    "    ax.text(1.4, 1.5, 'rho = %0.3f' % crispr_lr.rvalue, fontsize=pf.fontsize)\n",
    "    ax.text(1.4, 1.35, 'z = %0.2f' % crispr_z[geneA][geneB], fontsize=pf.fontsize)\n",
    "    pf.format_axis(ax)\n",
    "\n",
    "    # RNAi plot\n",
    "    plt.subplot(1, 2, 2)\n",
    "    rnai_df_filt = rnai_df[~pd.isnull(rnai_df[geneA]) & ~pd.isnull(rnai_df[geneB])]\n",
    "    ax = plt.gca()\n",
    "    ax.axhline(y=0, color='k', linewidth=lw)\n",
    "    ax.axvline(x=0, color='k', linewidth=lw)\n",
    "    plt.plot(rnai_df_filt[geneA].values, rnai_df_filt[geneB].values, marker='.', markersize=ms, linestyle='')\n",
    "    \n",
    "    # Plot linear regression\n",
    "    rnai_lr = linregress(rnai_df_filt[geneA].values, rnai_df_filt[geneB].values)\n",
    "    plt.plot(rnai_df_filt[geneA].values, rnai_df_filt[geneA].values * rnai_lr.slope + rnai_lr.intercept,\n",
    "             linestyle='-', linewidth=lw, color='black')\n",
    "    plt.xlabel(f'{geneA} Gene Effect (DEMETER2)')\n",
    "    plt.ylabel(f'{geneB} Gene Effect (DEMETER2)')\n",
    "    ax.text(0.7, 0.69, 'rho = %0.3f' % rnai_lr.rvalue, fontsize=pf.fontsize)\n",
    "    ax.text(0.7, 0.57, 'z = %0.2f' % rnai_z[geneA][geneB], fontsize=pf.fontsize)\n",
    "    fig.tight_layout()\n",
    "    pf.format_axis(ax)\n",
    "    plt.savefig(fig_path(f'{geneA}_{geneB}_corr_plot', 'pdf'))\n",
    "    \n",
    "plot_example('CDKN1A', 'CHEK2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure: Proportion of correlations between mito genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mito2_pcts = []\n",
    "mito1_pcts = []\n",
    "corr_range = np.linspace(0, 6, 13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def get_mito_pairs():\n",
    "    results = {}\n",
    "    for ix in range(len(corr_range)):\n",
    "        mito2_pairs = []\n",
    "        mito1_pairs = []\n",
    "        mito0_pairs = []\n",
    "        corr_lb = corr_range[ix]\n",
    "        corr_ub = None if (ix + 1) >= len(corr_range) \\\n",
    "                       else corr_range[ix + 1]\n",
    "        print(f\"Getting correlations between {corr_lb} and {corr_ub}\")\n",
    "        if corr_ub is None:\n",
    "            z_range = dep_z[(dep_z.abs() >= corr_lb)]\n",
    "        else:\n",
    "            z_range = dep_z[(dep_z.abs() >= corr_lb) & (dep_z.abs() < corr_ub)]\n",
    "        pair_iter = corr_matrix_to_generator(z_range, max_pairs=1000000)\n",
    "        print(\"Looping correlations\")\n",
    "        for (a, b), zsc in pair_iter:\n",
    "            # At least 1 mitogene\n",
    "            if (a in mitogenes) or (b in mitogenes):\n",
    "                mito1_pairs.append((a, b, zsc))\n",
    "                # 2 mitogenes\n",
    "                if (a in mitogenes) and (b in mitogenes):\n",
    "                    mito2_pairs.append((a, b, zsc))\n",
    "            # No mitogenes\n",
    "            else:\n",
    "                mito0_pairs.append((a, b, zsc))\n",
    "        results[(corr_lb, corr_ub)] = {'mito0': mito0_pairs,\n",
    "                                       'mito1': mito1_pairs,\n",
    "                                       'mito2': mito2_pairs}\n",
    "    return results\n",
    "\n",
    "mito_res = get_mito_pairs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mplayer\n",
    "p = mplayer.Player()\n",
    "success_file = '/Users/johnbachman/Downloads/success.m4a'\n",
    "p.loadfile(success_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_mito_pcts(mito_res):\n",
    "    m1_vals = []\n",
    "    m2_vals = []\n",
    "    lbs_vals = []\n",
    "    for corr_bounds, pair_dict in mito_res.items():\n",
    "        # Use the lower bound\n",
    "        lbs_vals.append(corr_bounds[0])\n",
    "        num_mito0 = len(pair_dict['mito0'])\n",
    "        num_mito1 = len(pair_dict['mito1'])\n",
    "        num_mito2 = len(pair_dict['mito2'])\n",
    "        total_corrs = num_mito0 + num_mito1\n",
    "        m1_vals.append(num_mito1 / total_corrs)\n",
    "        m2_vals.append(num_mito2 / total_corrs)\n",
    "        \n",
    "    plt.figure(figsize=(2, 2), dpi=150)\n",
    "    plt.plot(lbs_vals, m1_vals, color='b', label='1 mito gene')\n",
    "    plt.plot(lbs_vals, m2_vals, color='r', label='2 mito genes')\n",
    "    ax = plt.gca()\n",
    "    plt.xlabel('Absolute z-score lower bound')\n",
    "    plt.ylabel('Pct. of correlations')\n",
    "    plt.xticks(range(8))\n",
    "    plt.legend(loc='upper left', fontsize=pf.fontsize, frameon=False)\n",
    "    plt.subplots_adjust(left=0.16, bottom=0.15)\n",
    "    pf.format_axis(ax)\n",
    "    plt.savefig(fig_path(f'mito_corr_pcts', 'pdf'))\n",
    "    plt.savefig(fig_path(f'mito_corr_pcts', 'png'))\n",
    "    \n",
    "plot_mito_pcts(mito_res)   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hunting for other mitogenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def nonmito(cutoff):\n",
    "    nonmito_corrs = []\n",
    "    for (corr_lb, corr_ub), pairs_dict in mito_res.items():\n",
    "        mito1 = set(pairs_dict['mito1'])\n",
    "        mito2 = set(pairs_dict['mito2'])\n",
    "        only1 = mito1.difference(mito2)\n",
    "        for a, b, zsc in only1:\n",
    "            if abs(zsc) < 2:\n",
    "                continue\n",
    "            if a in mitogenes:\n",
    "                mitogene = a\n",
    "                nonmitogene = b\n",
    "            else:\n",
    "                mitogene = b\n",
    "                nonmitogene = a\n",
    "            nonmito_corrs.append((nonmitogene, mitogene, zsc))\n",
    "    nonmito_df = pd.DataFrame.from_records(nonmito_corrs, columns=['non_mito', 'mito', 'z_score'])\n",
    "    # Skips correlations involving only mito genes\n",
    "    # Find correlations with 1 mitogene\n",
    "    return nonmito_df\n",
    "nonmito_df3 = nonmito(3.0)\n",
    "gb = nonmito_df3.groupby('non_mito')\n",
    "mito_agg_mean = gb.aggregate(lambda x: x.abs().mean()).sort_values(by='z_score', ascending=False)\n",
    "mito_agg_mean[0:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on this analysis, MRPL58, and MCL1, TWNK (Twinkle mtDNA helicase), GATB (Glutamyl-TRNA(Gln) Amidotransferase Subunit B, Mitochondrial), MRM2 (Mitochondrial RRNA Methyltransferase 2), should definitely be included in mitocarta.\n",
    "\n",
    "COL4A3BP is a ceramide transporter/sphingolipid metabolism regulator, linked to SPTLC2 (sphingolipid metabolism, in Mitocarta). possible interesting link to mitos.\n",
    "\n",
    "WDR44 and ACAP2 both associated with endosomes/endocytic recycling. WDR44 also linked to iron uptake. Possible link to endosome/mitochondria relationship (which also involves iron). Interestingly, WDR44 is negatively correlated with many mitochondria ribosome subunits.\n",
    "\n",
    "Weirdly, WDR44 is in this dataset because of strong correlation with KIAA0100, which is listed as a mitochondrila gene? And ACAP2 included because linked to RAB35 (apparently mitochondrial).  Need to understand more about this.\n",
    "\n",
    "Iron transporter: GTPBP8 (ferrous iron transmembrane transporter activity), TFRC\n",
    "\n",
    "Nucleotide biosynthesis genes: ADSS (purine), GART (purine), PPAT (purine), ADSL (purine), PFAS (purine), UMPS (pyrimidine).\n",
    "\n",
    "Peroxisome:\n",
    "* PEX6. Peroxisomal biogenesis. Correlated with ACOX1 and HSD17B4, both in MitoCarta and with peroxisomal function annotations. Binds CERS2 (ceramide synthase), MARCH5.\n",
    "* PEX5, PEX12.\n",
    "\n",
    "Other known:\n",
    "* MCL1\n",
    "\n",
    "Mysterious:\n",
    "* LSM12. Binds PBP1 (stress granules, mRNA processing) and ATXN2, which is classified as a mitochondrial gene in Mitocarta (Genecards: \"The encoded cytoplasmic protein localizes to the endoplasmic reticulum and plasma membrane, is involved in endocytosis, and modulates mTOR signals, modifying ribosomal translation and mitochondrial function.\"). In Yeast, binds a bunch of RNA processing and ribosome proteins (https://thebiogrid.org/36554/summary/saccharomyces-cerevisiae/lsm12.html).\n",
    "* TAX1BP1. Correlated with NCOA4, androgen receptor coactivator classified as mitochondrial?\n",
    "* ZNF638. Correlated with TNFRSF10B (DR5), CFLAR, CASP8, BCL2L1 (Bcl-XL)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's aggregate by number of correlations (above cutoff) to see what comes up:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nonmito_df3[nonmito_df3.non_mito == 'TNFRSF10B']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "nonmito_df3 = nonmito(3.0)\n",
    "gb = nonmito_df3.groupby('non_mito')\n",
    "mito_agg_mean = gb.aggregate('count').sort_values(by='z_score', ascending=False)\n",
    "mito_agg_mean[0:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of these, some should be in Mitocarta:\n",
    "* KIAA0391 (\"Mitochondrial Ribonuclease P Catalytic Subunit\")\n",
    "* CLUH (Clustered Mitochondria Homolog)\n",
    "* MINOS1 (Mitochondrial Contact Site And Cristae Organizing System Subunit 10)\n",
    "\n",
    "Other:\n",
    "* GTPBP8 (iron)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'RHEBL1' in mitogenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame.from_records(nonmito_corrs[0:3], columns=['a', 'b', 'zsc'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure: For Mitogenes, strong correlations != stronger effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_mean = crispr_df.abs().mean(axis=0).sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_z_mean = crispr_z.abs().mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_corrs = 50\n",
    "np.mean(sorted(crispr_z['CHEK2'].abs(), reverse=True)[1:n_corrs+1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#crispr_z_mean = crispr_z.apply(lambda x: np.mean(sorted(x.abs(), reverse=True)[1:101]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_z_mean.CHEK2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "effect_corr = pd.concat([crispr_mean, crispr_z_mean], axis=1, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mito_effect = effect_corr[effect_corr.index.isin(mitogenes)]\n",
    "nonmito_effect = effect_corr[~effect_corr.index.isin(mitogenes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = nonmito_effect.groupby(pd.qcut(nonmito_effect[0].values, n_corrs)).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(nonmito_effect[0], nonmito_effect[1], linestyle='', marker='.', color='b', alpha=0.5, label='Not in MitoCarta')\n",
    "plt.plot(mito_effect[0], mito_effect[1], linestyle='', marker='.', color='r', alpha=0.5, label='In MitoCarta')\n",
    "plt.plot(res[0].values, res[1].values, linewidth=3, color='black')\n",
    "plt.xlabel('Average CRISPR Gene Effect')\n",
    "plt.ylabel(f'Avg of top {n_corrs} CRISPR Corrs (z-scores)')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Solute Carrier Family genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_df = pd.read_csv('data/slc_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_genes = list(slc_df['Approved symbol'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#inet_file = '/Users/johnbachman/Dropbox/1johndata/Knowledge File/Biology/Research' \\\n",
    "#            '/Big Mechanism/datasets/depmap_analysis/graphs/2021-01-06/indranet_dir_graph.pkl'\n",
    "#with open(inet_file, 'rb') as f:\n",
    "#    inet = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reactome dict\n",
    "reactome_file = '/Users/johnbachman/Dropbox/1johndata/Knowledge File/Biology/Research/Big Mechanism/' \\\n",
    "                'datasets/depmap_analysis/reactome_pathways.pkl'\n",
    "\n",
    "apriori_mapping = get_mitocarta_info(mito_file)\n",
    "\n",
    "with open(reactome_file, 'rb') as f:\n",
    "    up2path, _, pathid2pathname = pickle.load(f)\n",
    "    \n",
    "reactome_dict = {'uniprot_mapping': up2path,\n",
    "                 'pathid_name_mapping': pathid2pathname}\n",
    "\n",
    "expl_funcs=['apriori_explained', 'expl_ab', 'expl_ba', 'expl_axb',\n",
    "            'expl_bxa', 'find_cp', 'get_st', 'common_reactome_paths']\n",
    "\n",
    "sd_range = (2, None)\n",
    "dme_file = 'slc_explainer.pkl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_depmap(inet_file, z_filepath, dme_file, 'unsigned', sd_range, subset_list=slc_genes,\n",
    "           overwrite=True, reactome_path=reactome_file,\n",
    "           depmap_date='21q1', expl_funcs=expl_funcs, apriori_explained=mito_file, n_chunks=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mplayer\n",
    "p = mplayer.Player()\n",
    "success_file = '/Users/johnbachman/Downloads/success.m4a'\n",
    "p.loadfile(success_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(dme_file, 'rb') as f:\n",
    "    dme = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dme_pair = dme.stats_df.set_index('pair').drop(['agA', 'agB', 'z_score'], axis=1)\n",
    "dme_df = dme.expl_df.join(dme_pair, on='pair')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_to_expl = dme_df[(dme_df['z_score'] > 3) &\n",
    "                     (dme_df['apriori_explained'] == False) &\n",
    "                     (dme_df['reactome_paths'] == False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#axb_df = dme.get_filtered_triples_df()\n",
    "dme.get_filtered_triples_df?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dme_axb = dme.stats_df.set_index('pair').drop(['agA', 'agB', 'z_score', 'agA_ns', 'agA_id', 'agB_ns', 'agB_id'], axis=1)\n",
    "axb_all = axb_df.join(dme_axb, on='pair')\n",
    "axb_all['avg_x_corr'] = (axb_all.ax_corr.abs() + axb_all.bx_corr.abs()) / 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "axb_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_to_expl = axb_all[(axb_all['z_score'] > 3) &\n",
    "                         (axb_all['apriori_explained'] == False) &\n",
    "                         (axb_all['reactome_paths'] == False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_to_expl[slc_to_expl['agA'] ==]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_to_expl[slc_to_expl['avg_x_corr'] > 1.5].sort_values('avg_x_corr', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mplayer\n",
    "p = mplayer.Player()\n",
    "success_file = '/Users/johnbachman/Downloads/success.m4a'\n",
    "p.loadfile(success_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dme.plot_corr_stats('slc_output', z_corr=dep_z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dme.plot_dists('slc_output', z_corr=dep_z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mplayer\n",
    "p = mplayer.Player()\n",
    "success_file = '/Users/johnbachman/Downloads/success.m4a'\n",
    "p.loadfile(success_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene = 'SLC2A4'\n",
    "gene_df = slc_to_expl[(slc_to_expl['agA'] == gene) | (slc_to_expl['agB'] == gene)]\n",
    "expl_set = []\n",
    "for row in gene_df.itertuples():\n",
    "    if row.expl_type == 'shared_regulator':\n",
    "        continue\n",
    "    elif row.expl_type == 'shared_target':\n",
    "        expl_data = row.expl_data[2]\n",
    "    else:\n",
    "        expl_data = row.expl_data\n",
    "    for expl in expl_data:\n",
    "        expl_set.append(expl)\n",
    "expl_ctr = Counter(expl_set)\n",
    "expl_ctr = sorted([(k, v) for k, v in expl_ctr.items()], key=lambda x: x[1], reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df[0:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from indra.util import batch_iter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goi = 'MDM2'\n",
    "res = sort_corrs(dep_z, goi)\n",
    "#for r in res[0:20]:\n",
    "#    print(r)\n",
    "expl_pct = []\n",
    "batch_size = 1000\n",
    "for batch in batch_iter(res, batch_size):\n",
    "    counter = 0\n",
    "    for gene, corr in batch:\n",
    "        if gene in indra_dg[goi]:\n",
    "            counter += 1\n",
    "    pct = 100 * (counter / batch_size)\n",
    "    expl_pct.append(pct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.bar(range(len(expl_pct)), expl_pct)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Drug-Drug Correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug2_corrs = drug_resp_df.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from indra.sources import tas\n",
    "tp = tas.process_from_web()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tas_stmts = tp.statements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build up a dictionary of mappings from CHEBI to SMILES and INCHIKEY\n",
    "import pyobo\n",
    "chebi_inchikey_property = 'http://purl.obolibrary.org/obo/chebi/inchikey'\n",
    "chebi_smiles_property = 'http://purl.obolibrary.org/obo/chebi/smiles'\n",
    "chebi_id_to_smiles = pyobo.get_filtered_properties_mapping('chebi', chebi_smiles_property)\n",
    "chebi_id_to_inchikey = pyobo.get_filtered_properties_mapping('chebi', chebi_inchikey_property)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "foo = [stmt for stmt in tas_stmts if stmt.agent_list()[0].name == 'vemurafenib']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "foo[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "foo[0].subj.db_refs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chembl_ids = [stmt.agent_list()[0].db_refs.get('CHEMBL') for stmt in tas_stmts]\n",
    "chembl_ids = list(set([i for i in chembl_ids if i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(chembl_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chembl_ids_str = ';'.join(chembl_ids[0:200])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "j = res.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "j['molecules'][0]['molecule_chembl_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from indra.util import batch_iter\n",
    "batch_size = 200\n",
    "chembl_url = 'https://www.ebi.ac.uk/chembl/api/data/molecule/set/'\n",
    "inchi_keys = []\n",
    "for ix, batch in enumerate(batch_iter(chembl_ids, batch_size)):\n",
    "    chembl_ids_str = ';'.join(list(batch))\n",
    "    print(ix)\n",
    "    res = requests.get(chembl_url + chembl_ids_str, headers={'Accept': 'application/json'})\n",
    "    if res.status_code != 200:\n",
    "        print(\"Error occurred\")\n",
    "        break\n",
    "    for mol in j['molecules']:\n",
    "        cid = mol['molecule_chembl_id']\n",
    "        ik = mol['molecule_structures']['standard_inchi_key']\n",
    "        inchi_keys.append((cid, ik))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "j = res.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inchi_to_chembl = {t[1]: t[0] for t in inchi_keys}\n",
    "chembl_to_inchi = {t[0]: t[1] for t in inchi_keys}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_drug_id(name):\n",
    "    drug_rec = drug_info_df[drug_info_df.name == name]\n",
    "    return drug_rec.index.to_list()[0]\n",
    "\n",
    "def get_drug_name(drug_id):\n",
    "    drug_rec = drug_info_df.loc[drug_id]\n",
    "    return drug_rec['name']\n",
    "\n",
    "def get_drug_targets(drug_id):\n",
    "    drug_rec = drug_info_df.loc[drug_id]\n",
    "    targets = drug_rec['target']\n",
    "    if pd.isna(targets):\n",
    "        return []\n",
    "    else:\n",
    "        return [t.strip() for t in targets.split(',')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dump all smiles strings from the DepMap data\n",
    "depmap_smiles = set()\n",
    "for _, smiles_str in drug_info_df.smiles.items():\n",
    "    if pd.isna(smiles_str):\n",
    "        continue\n",
    "    for sm in smiles_str.split(','):\n",
    "        sm = sm.strip()\n",
    "        if sm:\n",
    "            depmap_smiles.add(sm.strip())\n",
    "            \n",
    "depmap_smiles = list(demap_smiles)\n",
    "\n",
    "with open('depmap_smiles.txt', 'wt') as f:\n",
    "    for sm in depmap_smiles:\n",
    "        f.write('%s\\n' % sm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert from SMILES to INCHIKEY using obabel:\n",
    "# $ obabel -ismi depmap_smiles.txt -oinchikey > depmap_inchikey.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, read the INCHIKEY databack in and build up a mapping dictionary\n",
    "depmap_inchikey = []\n",
    "with open('depmap_inchikey.txt', 'rt') as f:\n",
    "    depmap_inchikey = [l.strip() for l in f.readlines()]\n",
    "depmap_smiles_to_ik = dict(zip(depmap_smiles, depmap_inchikey))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "chembl_url = 'https://www.ebi.ac.uk/chembl/api/data/substructure/'\n",
    "dmik_to_chembl = {}\n",
    "for ix, ik in enumerate(depmap_smiles_to_ik.values()):\n",
    "    if ix > 10:\n",
    "        break\n",
    "    res = requests.get(chembl_url + ik, headers={'Accept': 'application/json'})\n",
    "    if res.status_code != 200:\n",
    "        print(\"Error occurred\")\n",
    "        break\n",
    "    if res.json()['molecules']:\n",
    "        for mol in res.json()['molecules']:\n",
    "            cid = mol['molecule_chembl_id']\n",
    "            if ik not in dmik_to_chembl:\n",
    "                dmik_to_chembl[ik] = [cid]\n",
    "            else:\n",
    "                dmik_to_chembl[ik].append(cid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dmik_to_chembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_name = 'vemurafenib'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "drug_info_df[drug_info_df['name'] == 'vemurafenib']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sm = [s.strip() for s in drug_info_df[drug_info_df['name'] == 'vemurafenib']['smiles'].values[0].split(',')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in sm:\n",
    "    ik = depmap_smiles_to_ik[s]\n",
    "    print(ik)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chembl_to_inchi['CHEMBL1229517']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_name = 'abemaciclib'\n",
    "corrs = sort_corrs(drug2_corrs, get_drug_id(drug_name))\n",
    "results = []\n",
    "corr_k = 20\n",
    "for i, (corr_drug_id, corr_val) in enumerate(corrs):\n",
    "    if corr_val == 1.0:\n",
    "        continue\n",
    "    if i >= corr_k:\n",
    "        break\n",
    "    corr_drug_name = get_drug_name(corr_drug_id)\n",
    "    corr_targets = get_drug_targets(corr_drug_id)\n",
    "    for t in corr_targets:\n",
    "        results.append((corr_drug_name, corr_drug_id, corr_val, t))\n",
    "tgt_df = pd.DataFrame.from_records(results, columns=('drug_name', 'drug_id', 'corr', 'target'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tgt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tgt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tgt_corrs = tgt_df.groupby('target')['corr'].sum()\n",
    "tgt_counts = tgt_df.groupby('target')['drug_id'].count()\n",
    "tgt_corrs.sort_values(ascending=False, inplace=True)\n",
    "tgt_counts.sort_values(ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_tgts = 20\n",
    "top_corrs = tgt_corrs[0:num_tgts]\n",
    "labels = top_corrs.index.to_list()\n",
    "corrs = top_corrs.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.bar(range(len(corrs)), corrs, align='center', tick_label=labels)\n",
    "plt.xticks(rotation='vertical')\n",
    "plt.subplots_adjust(bottom=0.2)\n",
    "plt.xlabel('Nominal targets of top %d correlates' % corr_k)\n",
    "plt.ylabel('Count')\n",
    "plt.title('Targets of drugs correlated with %s' % drug_name)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from indra.util import plot_formatting as pf\n",
    "from scipy.stats import linregress\n",
    "\n",
    "geneA = 'CDKN1A'\n",
    "geneB = 'CHEK2'\n",
    "lw = 0.5\n",
    "ms = 1\n",
    "fig = plt.figure(figsize=(4, 2))\n",
    "# CRISPR plot\n",
    "plt.subplot(1, 2, 1)\n",
    "ax = plt.gca()\n",
    "ax.axhline(y=0, color='k', linewidth=lw)\n",
    "ax.axvline(x=0, color='k', linewidth=lw)\n",
    "plt.plot(crispr_df[geneA].values, crispr_df[geneB].values, marker='.', markersize=ms, linestyle='')\n",
    "# Plot linear regression\n",
    "crispr_lr = linregress(crispr_df[geneA].values, crispr_df[geneB].values)\n",
    "plt.plot(crispr_df[geneA].values, crispr_df[geneA].values * crispr_lr.slope + crispr_lr.intercept,\n",
    "         linestyle='-', linewidth=lw, color='black')\n",
    "plt.xlabel(f'{geneA} Gene Effect (CERES)')\n",
    "plt.ylabel(f'{geneB} Gene Effect (CERES)')\n",
    "ax.text(1.4, 1.5, 'rho = %0.3f' % crispr_lr.rvalue, fontsize=pf.fontsize)\n",
    "ax.text(1.4, 1.35, 'z = %0.2f' % crispr_z[geneA][geneB], fontsize=pf.fontsize)\n",
    "pf.format_axis(ax)\n",
    "# RNAi plot\n",
    "plt.subplot(1, 2, 2)\n",
    "rnai_df_filt = rnai_df[~pd.isnull(rnai_df[geneA]) & ~pd.isnull(rnai_df[geneB])]\n",
    "ax = plt.gca()\n",
    "ax.axhline(y=0, color='k', linewidth=lw)\n",
    "ax.axvline(x=0, color='k', linewidth=lw)\n",
    "plt.plot(rnai_df_filt[geneA].values, rnai_df_filt[geneB].values, marker='.', markersize=ms, linestyle='')\n",
    "# Plot linear regression\n",
    "rnai_lr = linregress(rnai_df_filt[geneA].values, rnai_df_filt[geneB].values)\n",
    "plt.plot(rnai_df_filt[geneA].values, rnai_df_filt[geneA].values * rnai_lr.slope + rnai_lr.intercept,\n",
    "         linestyle='-', linewidth=lw, color='black')\n",
    "plt.xlabel(f'{geneA} Gene Effect (DEMETER2)')\n",
    "plt.ylabel(f'{geneB} Gene Effect (DEMETER2)')\n",
    "ax.text(0.7, 0.69, 'rho = %0.3f' % rnai_lr.rvalue, fontsize=pf.fontsize)\n",
    "ax.text(0.7, 0.57, 'z = %0.2f' % rnai_z[geneA][geneB], fontsize=pf.fontsize)\n",
    "\n",
    "pf.format_axis(ax)\n",
    "fig.tight_layout()\n",
    "plt.savefig(join(figdir, f'{geneA}_{geneB}_corr_plot.pdf'))\n",
    "dep_z[geneA][geneB]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deprecated/Old"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Cosmic Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cos_df = pd.read_csv('data/Census_allThu Jun 13 20_37_30 2019.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cos_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cos_corrs = corrs_for_genes(cos_df['Gene Symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "5+5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_df.head()\n",
    "slc_corrs = corrs_for_genes(slc_df['Approved symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mitocarta = pd.read_excel('data/Human.MitoCarta2.0.xls', sheet_name=1)\n",
    "mitogenes = list(mitocarta.Symbol.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'PNPT1' in mitogenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_corrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_corrs_filt = []\n",
    "slc_corrs_seen = set()\n",
    "for corr in slc_corrs:\n",
    "    # Filter out correlations we've already seen\n",
    "    corr_set = frozenset(corr)\n",
    "    if corr_set in slc_corrs_seen:\n",
    "        continue\n",
    "    else:\n",
    "        slc_corrs_seen.add(corr_set)\n",
    "    # Filter out self-correlations\n",
    "    if corr[0] == corr[1]:\n",
    "        continue\n",
    "    # Filter out mitochondrial gene correlations\n",
    "    elif corr[0] in mitogenes or corr[1] in mitogenes:\n",
    "        continue\n",
    "    else:\n",
    "        slc_corrs_filt.append(corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_corrs_filt.sort(key=lambda x: x[2], reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "slc_corrs_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(slc_corrs_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "with open('slc_corrs_no_mito.csv', 'wt') as f:\n",
    "    csv_writer = csv.writer(f, delimiter=',')\n",
    "    csv_writer.writerows(slc_corrs_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = {'source': None,\n",
    "         'target': None,\n",
    "         'stmt_filter': ['conversion', 'fplx'],\n",
    "         'node_filter': ['hgnc', 'fplx', 'chebi', 'pubchem', 'go', 'mesh'],\n",
    "         'node_blacklist': [],\n",
    "         'edge_hash_blacklist': [],\n",
    "         'path_length': 1,\n",
    "         'sign': 'no_sign',\n",
    "         'weighted': False,\n",
    "         'bsco': 0.,\n",
    "         'direct_only': False,\n",
    "         'curated_db_only': False,         \n",
    "         'fplx_expand': False,\n",
    "         'simple': False,\n",
    "         'k_shortest': False}\n",
    "\n",
    "def get_expl_corrs(gene_pairs):\n",
    "    no_path = []\n",
    "    has_2path = []\n",
    "    has_3path = []\n",
    "    for ix, (gene_a, gene_b, z_score) in enumerate(gene_pairs):\n",
    "        for source, target in ((gene_a, gene_b), (gene_b, gene_a)):\n",
    "            print(\"%d getting paths for %s, %s\" % (ix, source, target))\n",
    "            query['source'] = source\n",
    "            query['target'] = target\n",
    "            query['path_length'] = 1\n",
    "            res = requests.post('http://127.0.0.1:5000/query/submit', json=query).json()\n",
    "            if '2' in res['result']['paths_by_node_count']:\n",
    "                has_2path.append((source, target, z_score))\n",
    "            else:\n",
    "                query['path_length'] = 2\n",
    "                res = requests.post('http://127.0.0.1:5000/query/submit', json=query).json()\n",
    "                if '3' in res['result']['paths_by_node_count']:\n",
    "                    has_3path.append((source, target, z_score))                \n",
    "                else:\n",
    "                    no_path.append((source, target, z_score))\n",
    "    return {'no_path': no_path, 'has_2path': has_2path, 'has_3path': has_3path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paths = get_expl_corrs(sorted(slc_corrs, key=lambda x: x[2], reverse=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paths['has_2path']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paths['has_3path']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paths['no_path']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query['source'] = 'GRSF1'\n",
    "query['target'] = 'SLC30A9'\n",
    "query['path_length'] = 2\n",
    "res = requests.post('http://127.0.0.1:5000/query/submit', json=query).json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load INDRA SIF DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stmts_df = pd.read_csv('data/stmts_by_pair_type.csv')\n",
    "# Filter to HGNC only\n",
    "stmts_df = stmts_df[(stmts_df['agA_ns'] == 'HGNC') & (stmts_df['agB_ns'] == 'HGNC')]\n",
    "# Remove namespace and identifier columns, leaving only name\n",
    "stmts_df = stmts_df[['agA_name', 'agB_name', 'stmt_type', 'evidence_count']]\n",
    "# Remove self-edges\n",
    "stmts_df = stmts_df[stmts_df['agA_name'] != stmts_df['agB_name']]\n",
    "# Useful bits of Pandas to know, ultimately not used here:\n",
    "# pd.crosstab([foo['agA_name'], foo['agB_name']], foo['stmt_type'], foo['evidence_count'],\n",
    "#             aggfunc='sum', dropna=False).fillna(0)\n",
    "# baz = bar[bar.apply(lambda x: x.name[0] != x.name[1], axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_dict = {}\n",
    "for row in stmts_df.values:\n",
    "    try:\n",
    "        raw_corr = dep_z[row[0]][row[1]]\n",
    "        \"\"\"\n",
    "        if int(raw_corr) == 0:\n",
    "            corr = 0\n",
    "        elif int(raw_corr) < 0:\n",
    "            corr = -1\n",
    "        elif int(raw_corr) > 0:\n",
    "            corr = 1\n",
    "        \"\"\"\n",
    "        corr = raw_corr\n",
    "    except KeyError:\n",
    "        continue\n",
    "    stmt_type = row[2]\n",
    "    gene_pair = (row[0], row[1])\n",
    "    if gene_pair not in x_dict:\n",
    "        x_dict[gene_pair] = {'Corr': corr}\n",
    "    x_dict[gene_pair][stmt_type] = row[3]\n",
    "features = ['Activation', 'Inhibition', 'Corr']\n",
    "all_data = np.zeros((len(x_dict), len(features)))\n",
    "for row_ix, feat_dict in enumerate(x_dict.values()):\n",
    "    for feat_name, feat_val in feat_dict.items():\n",
    "        if feat_name not in features:\n",
    "            continue\n",
    "        feat_ix = features.index(feat_name)\n",
    "        all_data[row_ix, feat_ix] = feat_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "# Shuffle the matrix in place\n",
    "np.random.shuffle(all_data)\n",
    "# Balance the data\n",
    "ctr = Counter(all_data[:, -1])\n",
    "classes = sorted([(k, v) for k,  v in ctr.items()], key=lambda x: x[0])\n",
    "# Get the class with the fewest members\n",
    "min_class_count = min([t[1] for t in classes])\n",
    "min_class_count\n",
    "data_bal = np.zeros((min_class_count * len(classes), len(features)))\n",
    "row_ix = 0\n",
    "for one_class, count in classes:\n",
    "    num_class_rows = 0\n",
    "    for ix in range(all_data.shape[0]):\n",
    "        if all_data[ix, -1] == one_class:\n",
    "            data_bal[row_ix] = all_data[ix,:]\n",
    "            row_ix += 1\n",
    "            num_class_rows += 1\n",
    "        if num_class_rows >= min_class_count:\n",
    "            print(\"Finished class %s, at rows %s\" % (one_class, row_ix))\n",
    "            break\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Divide training and test\n",
    "np.random.shuffle(all_data)\n",
    "partition_ix = int(len(all_data) * 0.8)\n",
    "train_data = all_data[0:partition_ix, :]\n",
    "test_data = all_data[partition_ix:, :]\n",
    "x_train = train_data[:,:-1]\n",
    "y_train = train_data[:, -1]\n",
    "x_test = test_data[:, :-1]\n",
    "y_test = test_data[:, -1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multinomial Naive Bayes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from sklearn.naive_bayes import MultinomialNB\n",
    "nb = MultinomialNB()\n",
    "nb.fit(x_train, y_train)\n",
    "nb.score(x_test, y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "yctr = Counter(y_train)\n",
    "yclasses = sorted([(k, v) for k,  v in yctr.items()], key=lambda x: x[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[features + ['Count']] + list(zip(nb.coef_, classes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linear Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.metrics import explained_variance_score\n",
    "\n",
    "lr = LinearRegression()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lr.fit(x_train, y_train)\n",
    "lr.coef_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = lr.predict(x_train)\n",
    "explained_variance_score(y_train, y_pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_mean = np.full(y_train.shape, y_train.mean())\n",
    "explained_variance_score(y_train, y_mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred_test = lr.predict(x_test)\n",
    "explained_variance_score(y_test, y_pred_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load INDRA Statements (deprecated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload = False\n",
    "if reload:\n",
    "    inc_stmts = ac.load_statements('increase_amt.pkl')\n",
    "    dec_stmts = ac.load_statements('decrease_amt.pkl')\n",
    "    stmts = inc_stmts + dec_stmts\n",
    "    stmts = ac.map_grounding(stmts)\n",
    "    stmts = [s for s in stmts if s.subj and s.subj.db_refs.get('HGNC')\n",
    "                             and s.obj and s.obj.db_refs.get('HGNC')]\n",
    "    stmts = ac.run_preassembly(stmts)\n",
    "    ac.dump_statements(stmts, 'assembled_stmts.pkl')\n",
    "else:\n",
    "    stmts = ac.load_statements('assembled_stmts.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stmts_by_obj = defaultdict(list)\n",
    "for s in stmts:\n",
    "    gene_id = s.obj.db_refs['HGNC']\n",
    "    gene_name = hgnc_client.get_hgnc_name(gene_id)\n",
    "    stmts_by_obj[gene_name].append(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stmt_counts = []\n",
    "for obj, stmts in stmts_by_obj.items():\n",
    "    stmt_counts.append((obj, len(stmts)))\n",
    "stmt_counts.sort(key=lambda x: x[1], reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stmt_counts[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'BCL2'\n",
    "rank = 0\n",
    "stmts_by_obj[gene].sort(key=lambda s: s.belief, reverse=True)\n",
    "print(stmts_by_obj[gene][rank], '\\n')\n",
    "print('\\n'.join(['%d: %s\\n' % (i, str((e.source_api, e.text)))\n",
    "                 for i, e in enumerate(stmts_by_obj[gene][rank].evidence)]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking the number of infinities in the z scores:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18121"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(crispr_z.abs() == np.inf).sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "129863"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(crispr_z_ppf.abs() == np.inf).sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17313"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(rnai_z.abs() == np.inf).sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66249"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(rnai_z_ppf.abs() == np.inf).sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cases of strongly conflicting sign of correlation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are some interesting cases where nans appear in the dataset because the z-scores from small pvalues yield opposing +/- inf in CRISPR vs. RNAi. Most of these appear to be members of families."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], dtype: float64)"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val = 123456\n",
    "dep_z_copy = dep_z.copy()\n",
    "# Replace NaNs with an arbitrary value so we can get them later\n",
    "dep_z_copy[pd.isna(dep_z_copy)] = val\n",
    "confs = dep_z_copy[dep_z_copy == val].where(\n",
    "                np.triu(np.ones(dep_z_copy.shape), k=1).astype(np.bool))\n",
    "stacked: pd.DataFrame = confs.stack(dropna=True)\n",
    "stacked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's another weird case, where PSMA2 and RPL18 have such similarly (but not identical) extreme p-values that they are converted to identical z-scores and cancel out to 0 when added:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], dtype: float64)"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dep_z_copy = dep_z.copy()\n",
    "# Replace NaNs with an arbitrary value so we can get them later\n",
    "confs = dep_z_copy[dep_z_copy == 0].where(\n",
    "                np.triu(np.ones(dep_z_copy.shape), k=1).astype(np.bool))\n",
    "stacked: pd.DataFrame = confs.stack(dropna=True)\n",
    "stacked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-36.60307091042653"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_logp['PSMA2']['RPL18']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8.276444151723624"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_z['PSMA2']['RPL18']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rnai_logp' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mNameError\u001B[0m                                 Traceback (most recent call last)",
      "\u001B[0;32m<ipython-input-140-7ac0cadc5a25>\u001B[0m in \u001B[0;36m<module>\u001B[0;34m\u001B[0m\n\u001B[0;32m----> 1\u001B[0;31m \u001B[0mrnai_logp\u001B[0m\u001B[0;34m[\u001B[0m\u001B[0;34m'PSMA2'\u001B[0m\u001B[0;34m]\u001B[0m\u001B[0;34m[\u001B[0m\u001B[0;34m'RPL18'\u001B[0m\u001B[0;34m]\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m",
      "\u001B[0;31mNameError\u001B[0m: name 'rnai_logp' is not defined"
     ]
    }
   ],
   "source": [
    "rnai_logp['PSMA2']['RPL18']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnai_z['PSMA2']['RPL18']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "del dep_z_copy"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}